// @brave/brave-ui/old/headings/style.js
import styled from 'styled-components';
import { setTheme } from '../../helpers';
const StyledSharedHeading = styled.span `
  box-sizing: border-box;
  font-family: inherit;
  color: inherit;
  -webkit-font-smoothing: antialiased;
  user-select: none;
  cursor: default;
`;
const StyledHeadingTitle = styled(StyledSharedHeading.withComponent('h1')) `
  margin: 0;
  font-weight: 400;
  white-space: nowrap;
  font-size: 28px;
  color: rgb(255, 80, 0);
`;
const StyledHeadingTitleLabel = styled.sup `
  box-sizing: border-box;
  color: #999;
  font-size: 15px;
`;
const StyledSectionHeading = styled(StyledSharedHeading.withComponent('h2')) `
  color: rgb(68, 68, 68);
  font-size: 20px;
  margin: 0 0 20px;
  font-weight: 400;
`;
const StyledFeatureHeading = styled(StyledSharedHeading.withComponent('h2')) `
  color: #444444;
  font-weight: normal;
  font-size: 14px;
  margin: 18px 0 8px;
  min-width: 160px;
`;
const StyledH1 = styled.h1 `
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  color: ${(p) => setTheme(p.customStyle, 'color')};
  font-weight: ${(p) => setTheme(p.customStyle, 'fontWeight') || '400'};
  font-family: ${(p) => setTheme(p.customStyle, 'fontFamily') || 'inherit'};
  margin: ${(p) => setTheme(p.customStyle, 'margin') || '14px 0 22px'};
  font-size: ${(p) => setTheme(p.customStyle, 'fontSize') || '30px'};
  text-align: ${(p) => setTheme(p.customStyle, 'textAlign') || 'left'};
  line-height: ${(p) => setTheme(p.customStyle, 'lineHeight') || '44px'};
`;
const StyledH2 = styled.h2 `
  /* TBD */
`;
const StyledH3 = styled.h3 `
  /* TBD */
`;
const StyledH4 = styled.h4 `
  /* TBD */
`;
const StyledH5 = styled.h5 `
  /* TBD */
`;
const StyledH6 = styled.h6 `
  /* TBD */
`;
export { StyledHeadingTitle, StyledHeadingTitleLabel, StyledSectionHeading, StyledFeatureHeading, StyledH1, StyledH2, StyledH3, StyledH4, StyledH5, StyledH6 };
