// @glif/filecoin-wallet-provider/dist/errors/errors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/* tslint:disable max-classes-per-file */
var WalletProviderError = /*#__PURE__*/function (_Error) {
  (0, _inherits2["default"])(WalletProviderError, _Error);

  var _super = _createSuper(WalletProviderError);

  function WalletProviderError(args) {
    var _this;

    (0, _classCallCheck2["default"])(this, WalletProviderError);
    _this = _super.call(this, args.message);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "code", void 0);
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this), WalletProviderError.prototype);
    if (args.code) _this.code = args.code;
    return _this;
  }

  (0, _createClass2["default"])(WalletProviderError, [{
    key: "toJSON",
    value: function toJSON() {
      var res = {
        message: this.message
      };
      if (this.code) res.code = this.code;
      return res;
    }
  }]);
  return WalletProviderError;
}( /*#__PURE__*/(0, _wrapNativeSuper2["default"])(Error));

var LedgerLostConnectionError = /*#__PURE__*/function (_WalletProviderError) {
  (0, _inherits2["default"])(LedgerLostConnectionError, _WalletProviderError);

  var _super2 = _createSuper(LedgerLostConnectionError);

  function LedgerLostConnectionError() {
    var _this2;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref);
    (0, _classCallCheck2["default"])(this, LedgerLostConnectionError);
    _this2 = _super2.call(this, _objectSpread({
      message: args.message || 'connection lost'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this2), LedgerLostConnectionError.prototype);
    return _this2;
  }

  return LedgerLostConnectionError;
}(WalletProviderError);

var TransportNotSupportedError = /*#__PURE__*/function (_WalletProviderError2) {
  (0, _inherits2["default"])(TransportNotSupportedError, _WalletProviderError2);

  var _super3 = _createSuper(TransportNotSupportedError);

  function TransportNotSupportedError() {
    var _this3;

    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref2);
    (0, _classCallCheck2["default"])(this, TransportNotSupportedError);
    _this3 = _super3.call(this, _objectSpread({
      message: args.message || 'transport not supported by device'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this3), TransportNotSupportedError.prototype);
    return _this3;
  }

  return TransportNotSupportedError;
}(WalletProviderError);

var LedgerReplugError = /*#__PURE__*/function (_WalletProviderError3) {
  (0, _inherits2["default"])(LedgerReplugError, _WalletProviderError3);

  var _super4 = _createSuper(LedgerReplugError);

  function LedgerReplugError() {
    var _this4;

    var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref3);
    (0, _classCallCheck2["default"])(this, LedgerReplugError);
    _this4 = _super4.call(this, _objectSpread({
      message: args.message || 'unknown'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this4), LedgerReplugError.prototype);
    return _this4;
  }

  return LedgerReplugError;
}(WalletProviderError);

var LedgerDisconnectedError = /*#__PURE__*/function (_WalletProviderError4) {
  (0, _inherits2["default"])(LedgerDisconnectedError, _WalletProviderError4);

  var _super5 = _createSuper(LedgerDisconnectedError);

  function LedgerDisconnectedError() {
    var _this5;

    var _ref4 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref4);
    (0, _classCallCheck2["default"])(this, LedgerDisconnectedError);
    _this5 = _super5.call(this, _objectSpread({
      message: args.message || 'ledger device disconnected'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this5), LedgerDisconnectedError.prototype);
    return _this5;
  }

  return LedgerDisconnectedError;
}(WalletProviderError);

var LedgerInUseByAnotherApp = /*#__PURE__*/function (_WalletProviderError5) {
  (0, _inherits2["default"])(LedgerInUseByAnotherApp, _WalletProviderError5);

  var _super6 = _createSuper(LedgerInUseByAnotherApp);

  function LedgerInUseByAnotherApp() {
    var _this6;

    var _ref5 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref5);
    (0, _classCallCheck2["default"])(this, LedgerInUseByAnotherApp);
    _this6 = _super6.call(this, _objectSpread({
      message: args.message || 'ledger device in use by another app'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this6), LedgerInUseByAnotherApp.prototype);
    return _this6;
  }

  return LedgerInUseByAnotherApp;
}(WalletProviderError);

var LedgerNotFoundError = /*#__PURE__*/function (_WalletProviderError6) {
  (0, _inherits2["default"])(LedgerNotFoundError, _WalletProviderError6);

  var _super7 = _createSuper(LedgerNotFoundError);

  function LedgerNotFoundError() {
    var _this7;

    var _ref6 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref6);
    (0, _classCallCheck2["default"])(this, LedgerNotFoundError);
    _this7 = _super7.call(this, _objectSpread({
      message: args.message || 'ledger device not found'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this7), LedgerNotFoundError.prototype);
    return _this7;
  }

  return LedgerNotFoundError;
}(WalletProviderError);

var LedgerDeviceLockedError = /*#__PURE__*/function (_WalletProviderError7) {
  (0, _inherits2["default"])(LedgerDeviceLockedError, _WalletProviderError7);

  var _super8 = _createSuper(LedgerDeviceLockedError);

  function LedgerDeviceLockedError() {
    var _this8;

    var _ref7 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref7);
    (0, _classCallCheck2["default"])(this, LedgerDeviceLockedError);
    _this8 = _super8.call(this, _objectSpread({
      message: args.message || 'ledger device locked'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this8), LedgerDeviceLockedError.prototype);
    return _this8;
  }

  return LedgerDeviceLockedError;
}(WalletProviderError);

var LedgerFilecoinAppBadVersionError = /*#__PURE__*/function (_WalletProviderError8) {
  (0, _inherits2["default"])(LedgerFilecoinAppBadVersionError, _WalletProviderError8);

  var _super9 = _createSuper(LedgerFilecoinAppBadVersionError);

  function LedgerFilecoinAppBadVersionError() {
    var _this9;

    var _ref8 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref8);
    (0, _classCallCheck2["default"])(this, LedgerFilecoinAppBadVersionError);
    _this9 = _super9.call(this, _objectSpread({
      message: args.message || 'bad filecoin app version on ledger device'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this9), LedgerFilecoinAppBadVersionError.prototype);
    return _this9;
  }

  return LedgerFilecoinAppBadVersionError;
}(WalletProviderError);

var LedgerFilecoinAppNotOpenError = /*#__PURE__*/function (_WalletProviderError9) {
  (0, _inherits2["default"])(LedgerFilecoinAppNotOpenError, _WalletProviderError9);

  var _super10 = _createSuper(LedgerFilecoinAppNotOpenError);

  function LedgerFilecoinAppNotOpenError() {
    var _this10;

    var _ref9 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref9);
    (0, _classCallCheck2["default"])(this, LedgerFilecoinAppNotOpenError);
    _this10 = _super10.call(this, _objectSpread({
      message: args.message || 'filecoin app not open on device'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this10), LedgerFilecoinAppNotOpenError.prototype);
    return _this10;
  }

  return LedgerFilecoinAppNotOpenError;
}(WalletProviderError);

var LedgerDeviceBusyError = /*#__PURE__*/function (_WalletProviderError10) {
  (0, _inherits2["default"])(LedgerDeviceBusyError, _WalletProviderError10);

  var _super11 = _createSuper(LedgerDeviceBusyError);

  function LedgerDeviceBusyError() {
    var _this11;

    var _ref10 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref10);
    (0, _classCallCheck2["default"])(this, LedgerDeviceBusyError);
    _this11 = _super11.call(this, _objectSpread({
      message: args.message || 'ledger device locked or busy'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this11), LedgerDeviceBusyError.prototype);
    return _this11;
  }

  return LedgerDeviceBusyError;
}(WalletProviderError);

var InvalidParamsError = /*#__PURE__*/function (_WalletProviderError11) {
  (0, _inherits2["default"])(InvalidParamsError, _WalletProviderError11);

  var _super12 = _createSuper(InvalidParamsError);

  function InvalidParamsError() {
    var _this12;

    var _ref11 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref11);
    (0, _classCallCheck2["default"])(this, InvalidParamsError);
    _this12 = _super12.call(this, _objectSpread({
      message: args.message || 'invalid params'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this12), InvalidParamsError.prototype);
    return _this12;
  }

  return InvalidParamsError;
}(WalletProviderError);

var UnsupportedKeyTypeError = /*#__PURE__*/function (_WalletProviderError12) {
  (0, _inherits2["default"])(UnsupportedKeyTypeError, _WalletProviderError12);

  var _super13 = _createSuper(UnsupportedKeyTypeError);

  function UnsupportedKeyTypeError() {
    var _this13;

    var _ref12 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref12);
    (0, _classCallCheck2["default"])(this, UnsupportedKeyTypeError);
    _this13 = _super13.call(this, _objectSpread({
      message: args.message || 'unsupported key type'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this13), UnsupportedKeyTypeError.prototype);
    return _this13;
  }

  return UnsupportedKeyTypeError;
}(WalletProviderError);

var TransactionRejectedError = /*#__PURE__*/function (_WalletProviderError13) {
  (0, _inherits2["default"])(TransactionRejectedError, _WalletProviderError13);

  var _super14 = _createSuper(TransactionRejectedError);

  function TransactionRejectedError() {
    var _this14;

    var _ref13 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref13);
    (0, _classCallCheck2["default"])(this, TransactionRejectedError);
    _this14 = _super14.call(this, _objectSpread({
      message: args.message || 'Transaction rejected'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this14), TransactionRejectedError.prototype);
    return _this14;
  }

  return TransactionRejectedError;
}(WalletProviderError);

var MetaMaskError = /*#__PURE__*/function (_WalletProviderError14) {
  (0, _inherits2["default"])(MetaMaskError, _WalletProviderError14);

  var _super15 = _createSuper(MetaMaskError);

  function MetaMaskError() {
    var _this15;

    var _ref14 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref14);
    (0, _classCallCheck2["default"])(this, MetaMaskError);
    _this15 = _super15.call(this, _objectSpread({
      message: args.message || 'MetaMask error'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this15), MetaMaskError.prototype);
    return _this15;
  }

  return MetaMaskError;
}(WalletProviderError);

var MetaMaskNotInstalledError = /*#__PURE__*/function (_WalletProviderError15) {
  (0, _inherits2["default"])(MetaMaskNotInstalledError, _WalletProviderError15);

  var _super16 = _createSuper(MetaMaskNotInstalledError);

  function MetaMaskNotInstalledError() {
    var _this16;

    var _ref15 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref15);
    (0, _classCallCheck2["default"])(this, MetaMaskNotInstalledError);
    _this16 = _super16.call(this, _objectSpread({
      message: args.message || 'MetaMask not installed'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this16), MetaMaskNotInstalledError.prototype);
    return _this16;
  }

  return MetaMaskNotInstalledError;
}(WalletProviderError);

var MetaMaskSnapsNotSupportedError = /*#__PURE__*/function (_WalletProviderError16) {
  (0, _inherits2["default"])(MetaMaskSnapsNotSupportedError, _WalletProviderError16);

  var _super17 = _createSuper(MetaMaskSnapsNotSupportedError);

  function MetaMaskSnapsNotSupportedError() {
    var _this17;

    var _ref16 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref16);
    (0, _classCallCheck2["default"])(this, MetaMaskSnapsNotSupportedError);
    _this17 = _super17.call(this, _objectSpread({
      message: args.message || 'The installed version of MetaMask does not support Snaps'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this17), MetaMaskSnapsNotSupportedError.prototype);
    return _this17;
  }

  return MetaMaskSnapsNotSupportedError;
}(WalletProviderError);

var MetaMaskLockedError = /*#__PURE__*/function (_WalletProviderError17) {
  (0, _inherits2["default"])(MetaMaskLockedError, _WalletProviderError17);

  var _super18 = _createSuper(MetaMaskLockedError);

  function MetaMaskLockedError() {
    var _this18;

    var _ref17 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref17);
    (0, _classCallCheck2["default"])(this, MetaMaskLockedError);
    _this18 = _super18.call(this, _objectSpread({
      message: args.message || 'MetaMask locked'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this18), MetaMaskLockedError.prototype);
    return _this18;
  }

  return MetaMaskLockedError;
}(WalletProviderError);

var MetaMaskFilSnapNotInstalledError = /*#__PURE__*/function (_WalletProviderError18) {
  (0, _inherits2["default"])(MetaMaskFilSnapNotInstalledError, _WalletProviderError18);

  var _super19 = _createSuper(MetaMaskFilSnapNotInstalledError);

  function MetaMaskFilSnapNotInstalledError() {
    var _this19;

    var _ref18 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = (0, _extends2["default"])({}, _ref18);
    (0, _classCallCheck2["default"])(this, MetaMaskFilSnapNotInstalledError);
    _this19 = _super19.call(this, _objectSpread({
      message: args.message || 'FILSnap not installed'
    }, args));
    Object.setPrototypeOf((0, _assertThisInitialized2["default"])(_this19), MetaMaskFilSnapNotInstalledError.prototype);
    return _this19;
  }

  return MetaMaskFilSnapNotInstalledError;
}(WalletProviderError);

var _default = {
  InvalidParamsError: InvalidParamsError,
  LedgerLostConnectionError: LedgerLostConnectionError,
  LedgerReplugError: LedgerReplugError,
  LedgerDisconnectedError: LedgerDisconnectedError,
  LedgerInUseByAnotherApp: LedgerInUseByAnotherApp,
  LedgerNotFoundError: LedgerNotFoundError,
  LedgerDeviceLockedError: LedgerDeviceLockedError,
  LedgerFilecoinAppBadVersionError: LedgerFilecoinAppBadVersionError,
  LedgerFilecoinAppNotOpenError: LedgerFilecoinAppNotOpenError,
  LedgerDeviceBusyError: LedgerDeviceBusyError,
  MetaMaskError: MetaMaskError,
  MetaMaskNotInstalledError: MetaMaskNotInstalledError,
  MetaMaskSnapsNotSupportedError: MetaMaskSnapsNotSupportedError,
  MetaMaskLockedError: MetaMaskLockedError,
  MetaMaskFilSnapNotInstalledError: MetaMaskFilSnapNotInstalledError,
  TransactionRejectedError: TransactionRejectedError,
  TransportNotSupportedError: TransportNotSupportedError,
  UnsupportedKeyTypeError: UnsupportedKeyTypeError,
  WalletProviderError: WalletProviderError
};
exports["default"] = _default;
