// @brave/wallet-standard-brave/lib/cjs/account.js
"use strict";
// This is copied with modification from @wallet-standard/wallet
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _BraveWalletWalletAccount_address, _BraveWalletWalletAccount_publicKey, _BraveWalletWalletAccount_chains, _BraveWalletWalletAccount_features, _BraveWalletWalletAccount_label, _BraveWalletWalletAccount_icon;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BraveWalletWalletAccount = void 0;
const solana_js_1 = require("./solana.js");
const chains = solana_js_1.SOLANA_CHAINS;
const features = ['solana:signAndSendTransaction', 'solana:signMessage', 'solana:signTransaction'];
class BraveWalletWalletAccount {
    get address() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_address, "f");
    }
    get publicKey() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_publicKey, "f").slice();
    }
    get chains() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_chains, "f").slice();
    }
    get features() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_features, "f").slice();
    }
    get label() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_label, "f");
    }
    get icon() {
        return __classPrivateFieldGet(this, _BraveWalletWalletAccount_icon, "f");
    }
    constructor({ address, publicKey, label, icon }) {
        _BraveWalletWalletAccount_address.set(this, void 0);
        _BraveWalletWalletAccount_publicKey.set(this, void 0);
        _BraveWalletWalletAccount_chains.set(this, void 0);
        _BraveWalletWalletAccount_features.set(this, void 0);
        _BraveWalletWalletAccount_label.set(this, void 0);
        _BraveWalletWalletAccount_icon.set(this, void 0);
        if (new.target === BraveWalletWalletAccount) {
            Object.freeze(this);
        }
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_address, address, "f");
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_publicKey, publicKey, "f");
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_chains, chains, "f");
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_features, features, "f");
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_label, label, "f");
        __classPrivateFieldSet(this, _BraveWalletWalletAccount_icon, icon, "f");
    }
}
exports.BraveWalletWalletAccount = BraveWalletWalletAccount;
_BraveWalletWalletAccount_address = new WeakMap(), _BraveWalletWalletAccount_publicKey = new WeakMap(), _BraveWalletWalletAccount_chains = new WeakMap(), _BraveWalletWalletAccount_features = new WeakMap(), _BraveWalletWalletAccount_label = new WeakMap(), _BraveWalletWalletAccount_icon = new WeakMap();
