// @brave/wallet-standard-brave/lib/cjs/wallet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _BraveWalletWallet_instances, _BraveWalletWallet_listeners, _BraveWalletWallet_version, _BraveWalletWallet_name, _BraveWalletWallet_icon, _BraveWalletWallet_account, _BraveWalletWallet_braveWallet, _BraveWalletWallet_on, _BraveWalletWallet_emit, _BraveWalletWallet_off, _BraveWalletWallet_connected, _BraveWalletWallet_disconnected, _BraveWalletWallet_reconnected, _BraveWalletWallet_connect, _BraveWalletWallet_disconnect, _BraveWalletWallet_signAndSendTransaction, _BraveWalletWallet_signTransaction, _BraveWalletWallet_signMessage;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BraveWalletWallet = void 0;
const web3_js_1 = require("@solana/web3.js");
const bs58_1 = __importDefault(require("bs58"));
const account_js_1 = require("./account.js");
const icon_js_1 = require("./icon.js");
const solana_js_1 = require("./solana.js");
const util_js_1 = require("./util.js");
class BraveWalletWallet {
    get version() {
        return __classPrivateFieldGet(this, _BraveWalletWallet_version, "f");
    }
    get name() {
        return __classPrivateFieldGet(this, _BraveWalletWallet_name, "f");
    }
    get icon() {
        return __classPrivateFieldGet(this, _BraveWalletWallet_icon, "f");
    }
    get chains() {
        return solana_js_1.SOLANA_CHAINS.slice();
    }
    get features() {
        return {
            'standard:connect': {
                version: '1.0.0',
                connect: __classPrivateFieldGet(this, _BraveWalletWallet_connect, "f"),
            },
            'standard:disconnect': {
                version: '1.0.0',
                disconnect: __classPrivateFieldGet(this, _BraveWalletWallet_disconnect, "f"),
            },
            'standard:events': {
                version: '1.0.0',
                on: __classPrivateFieldGet(this, _BraveWalletWallet_on, "f"),
            },
            'solana:signAndSendTransaction': {
                version: '1.0.0',
                supportedTransactionVersions: ['legacy', 0],
                signAndSendTransaction: __classPrivateFieldGet(this, _BraveWalletWallet_signAndSendTransaction, "f"),
            },
            'solana:signTransaction': {
                version: '1.0.0',
                supportedTransactionVersions: ['legacy', 0],
                signTransaction: __classPrivateFieldGet(this, _BraveWalletWallet_signTransaction, "f"),
            },
            'solana:signMessage': {
                version: '1.0.0',
                signMessage: __classPrivateFieldGet(this, _BraveWalletWallet_signMessage, "f"),
            },
            'braveWallet:': {
                braveWallet: __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f"),
            },
        };
    }
    get accounts() {
        return __classPrivateFieldGet(this, _BraveWalletWallet_account, "f") ? [__classPrivateFieldGet(this, _BraveWalletWallet_account, "f")] : [];
    }
    constructor(braveWallet) {
        _BraveWalletWallet_instances.add(this);
        _BraveWalletWallet_listeners.set(this, {});
        _BraveWalletWallet_version.set(this, '1.0.0');
        _BraveWalletWallet_name.set(this, 'Brave Wallet');
        _BraveWalletWallet_icon.set(this, icon_js_1.icon);
        _BraveWalletWallet_account.set(this, null);
        _BraveWalletWallet_braveWallet.set(this, void 0);
        _BraveWalletWallet_on.set(this, (event, listener) => {
            var _a;
            ((_a = __classPrivateFieldGet(this, _BraveWalletWallet_listeners, "f")[event]) === null || _a === void 0 ? void 0 : _a.push(listener)) || (__classPrivateFieldGet(this, _BraveWalletWallet_listeners, "f")[event] = [listener]);
            return () => __classPrivateFieldGet(this, _BraveWalletWallet_instances, "m", _BraveWalletWallet_off).call(this, event, listener);
        });
        _BraveWalletWallet_connected.set(this, () => {
            var _a;
            const address = (_a = __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").publicKey) === null || _a === void 0 ? void 0 : _a.toBase58();
            if (address) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const publicKey = __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").publicKey.toBytes();
                const account = __classPrivateFieldGet(this, _BraveWalletWallet_account, "f");
                if (!account || account.address !== address || !(0, util_js_1.bytesEqual)(account.publicKey, publicKey)) {
                    __classPrivateFieldSet(this, _BraveWalletWallet_account, new account_js_1.BraveWalletWalletAccount({ address, publicKey }), "f");
                    __classPrivateFieldGet(this, _BraveWalletWallet_instances, "m", _BraveWalletWallet_emit).call(this, 'change', { accounts: this.accounts });
                }
            }
        });
        _BraveWalletWallet_disconnected.set(this, () => {
            if (__classPrivateFieldGet(this, _BraveWalletWallet_account, "f")) {
                __classPrivateFieldSet(this, _BraveWalletWallet_account, null, "f");
                __classPrivateFieldGet(this, _BraveWalletWallet_instances, "m", _BraveWalletWallet_emit).call(this, 'change', { accounts: this.accounts });
            }
        });
        _BraveWalletWallet_reconnected.set(this, () => {
            if (__classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").publicKey) {
                __classPrivateFieldGet(this, _BraveWalletWallet_connected, "f").call(this);
            }
            else {
                __classPrivateFieldGet(this, _BraveWalletWallet_disconnected, "f").call(this);
            }
        });
        _BraveWalletWallet_connect.set(this, ({ silent } = {}) => __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _BraveWalletWallet_account, "f")) {
                yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").connect(silent ? { onlyIfTrusted: true } : undefined);
            }
            __classPrivateFieldGet(this, _BraveWalletWallet_connected, "f").call(this);
            return { accounts: this.accounts };
        }));
        _BraveWalletWallet_disconnect.set(this, () => __awaiter(this, void 0, void 0, function* () {
            yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").disconnect();
        }));
        _BraveWalletWallet_signAndSendTransaction.set(this, (...inputs) => __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                throw new Error('not connected');
            const outputs = [];
            if (inputs.length === 1) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const { transaction, account, chain, options } = inputs[0];
                const { minContextSlot, preflightCommitment, skipPreflight, maxRetries } = options || {};
                if (account !== __classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                    throw new Error('invalid account');
                if (!(0, solana_js_1.isSolanaChain)(chain))
                    throw new Error('invalid chain');
                const { signature } = yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").signAndSendTransaction(web3_js_1.VersionedTransaction.deserialize(transaction), {
                    preflightCommitment,
                    minContextSlot,
                    maxRetries,
                    skipPreflight,
                });
                outputs.push({ signature: bs58_1.default.decode(signature) });
            }
            else if (inputs.length > 1) {
                for (const input of inputs) {
                    outputs.push(...(yield __classPrivateFieldGet(this, _BraveWalletWallet_signAndSendTransaction, "f").call(this, input)));
                }
            }
            return outputs;
        }));
        _BraveWalletWallet_signTransaction.set(this, (...inputs) => __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                throw new Error('not connected');
            const outputs = [];
            if (inputs.length === 1) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const { transaction, account, chain } = inputs[0];
                if (account !== __classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                    throw new Error('invalid account');
                if (chain && !(0, solana_js_1.isSolanaChain)(chain))
                    throw new Error('invalid chain');
                const signedTransaction = yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").signTransaction(web3_js_1.VersionedTransaction.deserialize(transaction));
                const serializedTransaction = (0, solana_js_1.isVersionedTransaction)(signedTransaction)
                    ? signedTransaction.serialize()
                    : new Uint8Array(signedTransaction.serialize({
                        requireAllSignatures: false,
                        verifySignatures: false,
                    }));
                outputs.push({ signedTransaction: serializedTransaction });
            }
            else if (inputs.length > 1) {
                let chain = undefined;
                for (const input of inputs) {
                    if (input.account !== __classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                        throw new Error('invalid account');
                    if (input.chain) {
                        if (!(0, solana_js_1.isSolanaChain)(input.chain))
                            throw new Error('invalid chain');
                        if (chain) {
                            if (input.chain !== chain)
                                throw new Error('conflicting chain');
                        }
                        else {
                            chain = input.chain;
                        }
                    }
                }
                const transactions = inputs.map(({ transaction }) => web3_js_1.VersionedTransaction.deserialize(transaction));
                const signedTransactions = yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").signAllTransactions(transactions);
                outputs.push(...signedTransactions.map((signedTransaction) => {
                    const serializedTransaction = (0, solana_js_1.isVersionedTransaction)(signedTransaction)
                        ? signedTransaction.serialize()
                        : new Uint8Array(signedTransaction.serialize({
                            requireAllSignatures: false,
                            verifySignatures: false,
                        }));
                    return { signedTransaction: serializedTransaction };
                }));
            }
            return outputs;
        }));
        _BraveWalletWallet_signMessage.set(this, (...inputs) => __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                throw new Error('not connected');
            const outputs = [];
            if (inputs.length === 1) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const { message, account } = inputs[0];
                if (account !== __classPrivateFieldGet(this, _BraveWalletWallet_account, "f"))
                    throw new Error('invalid account');
                const { signature } = yield __classPrivateFieldGet(this, _BraveWalletWallet_braveWallet, "f").signMessage(message);
                outputs.push({ signedMessage: message, signature });
            }
            else if (inputs.length > 1) {
                for (const input of inputs) {
                    outputs.push(...(yield __classPrivateFieldGet(this, _BraveWalletWallet_signMessage, "f").call(this, input)));
                }
            }
            return outputs;
        }));
        if (new.target === BraveWalletWallet) {
            Object.freeze(this);
        }
        __classPrivateFieldSet(this, _BraveWalletWallet_braveWallet, braveWallet, "f");
        braveWallet.on('connect', __classPrivateFieldGet(this, _BraveWalletWallet_connected, "f"), this);
        braveWallet.on('disconnect', __classPrivateFieldGet(this, _BraveWalletWallet_disconnected, "f"), this);
        braveWallet.on('accountChanged', __classPrivateFieldGet(this, _BraveWalletWallet_reconnected, "f"), this);
        __classPrivateFieldGet(this, _BraveWalletWallet_connected, "f").call(this);
    }
}
exports.BraveWalletWallet = BraveWalletWallet;
_BraveWalletWallet_listeners = new WeakMap(), _BraveWalletWallet_version = new WeakMap(), _BraveWalletWallet_name = new WeakMap(), _BraveWalletWallet_icon = new WeakMap(), _BraveWalletWallet_account = new WeakMap(), _BraveWalletWallet_braveWallet = new WeakMap(), _BraveWalletWallet_on = new WeakMap(), _BraveWalletWallet_connected = new WeakMap(), _BraveWalletWallet_disconnected = new WeakMap(), _BraveWalletWallet_reconnected = new WeakMap(), _BraveWalletWallet_connect = new WeakMap(), _BraveWalletWallet_disconnect = new WeakMap(), _BraveWalletWallet_signAndSendTransaction = new WeakMap(), _BraveWalletWallet_signTransaction = new WeakMap(), _BraveWalletWallet_signMessage = new WeakMap(), _BraveWalletWallet_instances = new WeakSet(), _BraveWalletWallet_emit = function _BraveWalletWallet_emit(event, ...args) {
    var _a;
    // eslint-disable-next-line prefer-spread
    (_a = __classPrivateFieldGet(this, _BraveWalletWallet_listeners, "f")[event]) === null || _a === void 0 ? void 0 : _a.forEach((listener) => listener.apply(null, args));
}, _BraveWalletWallet_off = function _BraveWalletWallet_off(event, listener) {
    var _a;
    __classPrivateFieldGet(this, _BraveWalletWallet_listeners, "f")[event] = (_a = __classPrivateFieldGet(this, _BraveWalletWallet_listeners, "f")[event]) === null || _a === void 0 ? void 0 : _a.filter((existingListener) => listener !== existingListener);
};
