// @brave/leo/tokens/css/variables.ts
/**
 * Do not edit directly
 * Generated on Thu Sep 19 2024 18:28:44 GMT+0000 (Coordinated Universal Time)
 */


export const color = {
  "toString": [
    ""
  ],
  "legacy": {
    "toString": [
      ""
    ],
    /** #ffffff */
    "background1": "var(--leo-color-legacy-background1)",
    /** #f8f9fa */
    "background2": "var(--leo-color-legacy-background2)",
    /** #212529 */
    "text1": "var(--leo-color-legacy-text1)",
    /** #495057 */
    "text2": "var(--leo-color-legacy-text2)",
    /** #868e96 */
    "text3": "var(--leo-color-legacy-text3)",
    /** #ea3a0d */
    "interactive1": "var(--leo-color-legacy-interactive1)",
    /** #fb542b */
    "interactive2": "var(--leo-color-legacy-interactive2)",
    /** #ff7654 */
    "interactive3": "var(--leo-color-legacy-interactive3)",
    /** #353dab */
    "interactive4": "var(--leo-color-legacy-interactive4)",
    /** #4c54d2 */
    "interactive5": "var(--leo-color-legacy-interactive5)",
    /** #737ade */
    "interactive6": "var(--leo-color-legacy-interactive6)",
    /** #212529 */
    "interactive7": "var(--leo-color-legacy-interactive7)",
    /** #aeb1c2 */
    "interactive8": "var(--leo-color-legacy-interactive8)",
    /** #dadce8 */
    "disabled": "var(--leo-color-legacy-disabled)",
    /** #a0a5eb */
    "focusBorder": "var(--leo-color-legacy-focus-border)",
    /** #e9e9f4 */
    "divider1": "var(--leo-color-legacy-divider1)"
  },
  "primitive": {
    "toString": [
      ""
    ],
    "neutral": {
      /** #000000 */
      "0": "var(--leo-color-primitive-neutral-0)",
      /** #111114 */
      "5": "var(--leo-color-primitive-neutral-5)",
      /** #1b1b1f */
      "10": "var(--leo-color-primitive-neutral-10)",
      /** #252529 */
      "15": "var(--leo-color-primitive-neutral-15)",
      /** #303034 */
      "20": "var(--leo-color-primitive-neutral-20)",
      /** #3b3b3f */
      "25": "var(--leo-color-primitive-neutral-25)",
      /** #46464a */
      "30": "var(--leo-color-primitive-neutral-30)",
      /** #525256 */
      "35": "var(--leo-color-primitive-neutral-35)",
      /** #5e5e62 */
      "40": "var(--leo-color-primitive-neutral-40)",
      /** #76767a */
      "50": "var(--leo-color-primitive-neutral-50)",
      /** #909094 */
      "60": "var(--leo-color-primitive-neutral-60)",
      /** #aaaaaf */
      "70": "var(--leo-color-primitive-neutral-70)",
      /** #c7c7cc */
      "80": "var(--leo-color-primitive-neutral-80)",
      /** #e3e3e8 */
      "90": "var(--leo-color-primitive-neutral-90)",
      /** #f0f0f4 */
      "95": "var(--leo-color-primitive-neutral-95)",
      /** #fafafb */
      "98": "var(--leo-color-primitive-neutral-98)",
      /** #fcfcfe */
      "99": "var(--leo-color-primitive-neutral-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-neutral-100)",
      "toString": [
        ""
      ]
    },
    "primary": {
      /** #000000 */
      "0": "var(--leo-color-primitive-primary-0)",
      /** #05004c */
      "5": "var(--leo-color-primitive-primary-5)",
      /** #0f0663 */
      "10": "var(--leo-color-primitive-primary-10)",
      /** #190b7d */
      "15": "var(--leo-color-primitive-primary-15)",
      /** #251491 */
      "20": "var(--leo-color-primitive-primary-20)",
      /** #2e19b0 */
      "25": "var(--leo-color-primitive-primary-25)",
      /** #372cbf */
      "30": "var(--leo-color-primitive-primary-30)",
      /** #3e37d4 */
      "35": "var(--leo-color-primitive-primary-35)",
      /** #4a46e0 */
      "40": "var(--leo-color-primitive-primary-40)",
      /** #6261ff */
      "50": "var(--leo-color-primitive-primary-50)",
      /** #8283ff */
      "60": "var(--leo-color-primitive-primary-60)",
      /** #a1a2ff */
      "70": "var(--leo-color-primitive-primary-70)",
      /** #c1c4ff */
      "80": "var(--leo-color-primitive-primary-80)",
      /** #dfe1ff */
      "90": "var(--leo-color-primitive-primary-90)",
      /** #efefff */
      "95": "var(--leo-color-primitive-primary-95)",
      /** #f8f8ff */
      "98": "var(--leo-color-primitive-primary-98)",
      /** #fbfbff */
      "99": "var(--leo-color-primitive-primary-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-primary-100)",
      "toString": [
        ""
      ]
    },
    "secondary": {
      /** #000000 */
      "0": "var(--leo-color-primitive-secondary-0)",
      /** #260600 */
      "5": "var(--leo-color-primitive-secondary-5)",
      /** #380c00 */
      "10": "var(--leo-color-primitive-secondary-10)",
      /** #4a1300 */
      "15": "var(--leo-color-primitive-secondary-15)",
      /** #5c1a00 */
      "20": "var(--leo-color-primitive-secondary-20)",
      /** #6f2000 */
      "25": "var(--leo-color-primitive-secondary-25)",
      /** #822800 */
      "30": "var(--leo-color-primitive-secondary-30)",
      /** #962f00 */
      "35": "var(--leo-color-primitive-secondary-35)",
      /** #aa3700 */
      "40": "var(--leo-color-primitive-secondary-40)",
      /** #d44600 */
      "50": "var(--leo-color-primitive-secondary-50)",
      /** #ff5705 */
      "60": "var(--leo-color-primitive-secondary-60)",
      /** #ff8b62 */
      "70": "var(--leo-color-primitive-secondary-70)",
      /** #ffb59c */
      "80": "var(--leo-color-primitive-secondary-80)",
      /** #ffdbcf */
      "90": "var(--leo-color-primitive-secondary-90)",
      /** #ffede8 */
      "95": "var(--leo-color-primitive-secondary-95)",
      /** #fff8f6 */
      "98": "var(--leo-color-primitive-secondary-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-secondary-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-secondary-100)",
      "toString": [
        ""
      ]
    },
    "tertiary": {
      /** #000000 */
      "0": "var(--leo-color-primitive-tertiary-0)",
      /** #2b0014 */
      "5": "var(--leo-color-primitive-tertiary-5)",
      /** #3e001f */
      "10": "var(--leo-color-primitive-tertiary-10)",
      /** #51002a */
      "15": "var(--leo-color-primitive-tertiary-15)",
      /** #640036 */
      "20": "var(--leo-color-primitive-tertiary-20)",
      /** #780042 */
      "25": "var(--leo-color-primitive-tertiary-25)",
      /** #8d004e */
      "30": "var(--leo-color-primitive-tertiary-30)",
      /** #a3005b */
      "35": "var(--leo-color-primitive-tertiary-35)",
      /** #b90068 */
      "40": "var(--leo-color-primitive-tertiary-40)",
      /** #e60083 */
      "50": "var(--leo-color-primitive-tertiary-50)",
      /** #ff479c */
      "60": "var(--leo-color-primitive-tertiary-60)",
      /** #ff83b2 */
      "70": "var(--leo-color-primitive-tertiary-70)",
      /** #ffb0ca */
      "80": "var(--leo-color-primitive-tertiary-80)",
      /** #ffd9e3 */
      "90": "var(--leo-color-primitive-tertiary-90)",
      /** #ffecf0 */
      "95": "var(--leo-color-primitive-tertiary-95)",
      /** #fff8f8 */
      "98": "var(--leo-color-primitive-tertiary-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-tertiary-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-tertiary-100)",
      "toString": [
        ""
      ]
    },
    "red": {
      /** #000000 */
      "0": "var(--leo-color-primitive-red-0)",
      /** #2c0004 */
      "5": "var(--leo-color-primitive-red-5)",
      /** #410008 */
      "10": "var(--leo-color-primitive-red-10)",
      /** #54000e */
      "15": "var(--leo-color-primitive-red-15)",
      /** #680014 */
      "20": "var(--leo-color-primitive-red-20)",
      /** #7d001a */
      "25": "var(--leo-color-primitive-red-25)",
      /** #920020 */
      "30": "var(--leo-color-primitive-red-30)",
      /** #a80026 */
      "35": "var(--leo-color-primitive-red-35)",
      /** #bf002c */
      "40": "var(--leo-color-primitive-red-40)",
      /** #e62440 */
      "50": "var(--leo-color-primitive-red-50)",
      /** #ff525e */
      "60": "var(--leo-color-primitive-red-60)",
      /** #ff888a */
      "70": "var(--leo-color-primitive-red-70)",
      /** #ffb3b2 */
      "80": "var(--leo-color-primitive-red-80)",
      /** #ffdad9 */
      "90": "var(--leo-color-primitive-red-90)",
      /** #ffedec */
      "95": "var(--leo-color-primitive-red-95)",
      /** #fff8f7 */
      "98": "var(--leo-color-primitive-red-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-red-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-red-100)",
      "toString": [
        ""
      ]
    },
    "orange": {
      /** #000000 */
      "0": "var(--leo-color-primitive-orange-0)",
      /** #260600 */
      "5": "var(--leo-color-primitive-orange-5)",
      /** #380c00 */
      "10": "var(--leo-color-primitive-orange-10)",
      /** #4a1300 */
      "15": "var(--leo-color-primitive-orange-15)",
      /** #5c1a00 */
      "20": "var(--leo-color-primitive-orange-20)",
      /** #6f2000 */
      "25": "var(--leo-color-primitive-orange-25)",
      /** #822800 */
      "30": "var(--leo-color-primitive-orange-30)",
      /** #962f00 */
      "35": "var(--leo-color-primitive-orange-35)",
      /** #aa3700 */
      "40": "var(--leo-color-primitive-orange-40)",
      /** #d44600 */
      "50": "var(--leo-color-primitive-orange-50)",
      /** #ff5705 */
      "60": "var(--leo-color-primitive-orange-60)",
      /** #ff8b62 */
      "70": "var(--leo-color-primitive-orange-70)",
      /** #ffb59c */
      "80": "var(--leo-color-primitive-orange-80)",
      /** #ffdbcf */
      "90": "var(--leo-color-primitive-orange-90)",
      /** #ffede8 */
      "95": "var(--leo-color-primitive-orange-95)",
      /** #fff8f6 */
      "98": "var(--leo-color-primitive-orange-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-orange-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-orange-100)",
      "toString": [
        ""
      ]
    },
    "yellow": {
      /** #000000 */
      "0": "var(--leo-color-primitive-yellow-0)",
      /** #171000 */
      "5": "var(--leo-color-primitive-yellow-5)",
      /** #241a00 */
      "10": "var(--leo-color-primitive-yellow-10)",
      /** #302400 */
      "15": "var(--leo-color-primitive-yellow-15)",
      /** #3d2f00 */
      "20": "var(--leo-color-primitive-yellow-20)",
      /** #4a3900 */
      "25": "var(--leo-color-primitive-yellow-25)",
      /** #584400 */
      "30": "var(--leo-color-primitive-yellow-30)",
      /** #665000 */
      "35": "var(--leo-color-primitive-yellow-35)",
      /** #745b00 */
      "40": "var(--leo-color-primitive-yellow-40)",
      /** #927300 */
      "50": "var(--leo-color-primitive-yellow-50)",
      /** #b18c00 */
      "60": "var(--leo-color-primitive-yellow-60)",
      /** #d1a600 */
      "70": "var(--leo-color-primitive-yellow-70)",
      /** #f1c00f */
      "80": "var(--leo-color-primitive-yellow-80)",
      /** #ffe08d */
      "90": "var(--leo-color-primitive-yellow-90)",
      /** #ffefce */
      "95": "var(--leo-color-primitive-yellow-95)",
      /** #fff8f1 */
      "98": "var(--leo-color-primitive-yellow-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-yellow-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-yellow-100)",
      "toString": [
        ""
      ]
    },
    "green": {
      /** #000000 */
      "0": "var(--leo-color-primitive-green-0)",
      /** #001501 */
      "5": "var(--leo-color-primitive-green-5)",
      /** #002202 */
      "10": "var(--leo-color-primitive-green-10)",
      /** #002d03 */
      "15": "var(--leo-color-primitive-green-15)",
      /** #003a05 */
      "20": "var(--leo-color-primitive-green-20)",
      /** #004607 */
      "25": "var(--leo-color-primitive-green-25)",
      /** #00530a */
      "30": "var(--leo-color-primitive-green-30)",
      /** #00610e */
      "35": "var(--leo-color-primitive-green-35)",
      /** #006e11 */
      "40": "var(--leo-color-primitive-green-40)",
      /** #008b18 */
      "50": "var(--leo-color-primitive-green-50)",
      /** #02a820 */
      "60": "var(--leo-color-primitive-green-60)",
      /** #38c43b */
      "70": "var(--leo-color-primitive-green-70)",
      /** #58e155 */
      "80": "var(--leo-color-primitive-green-80)",
      /** #b9f1ac */
      "90": "var(--leo-color-primitive-green-90)",
      /** #cbfbc0 */
      "95": "var(--leo-color-primitive-green-95)",
      /** #ecffe3 */
      "98": "var(--leo-color-primitive-green-98)",
      /** #f7ffef */
      "99": "var(--leo-color-primitive-green-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-green-100)",
      "toString": [
        ""
      ]
    },
    "teal": {
      /** #000000 */
      "0": "var(--leo-color-primitive-teal-0)",
      /** #001417 */
      "5": "var(--leo-color-primitive-teal-5)",
      /** #001f24 */
      "10": "var(--leo-color-primitive-teal-10)",
      /** #002a30 */
      "15": "var(--leo-color-primitive-teal-15)",
      /** #00363d */
      "20": "var(--leo-color-primitive-teal-20)",
      /** #00424b */
      "25": "var(--leo-color-primitive-teal-25)",
      /** #004f58 */
      "30": "var(--leo-color-primitive-teal-30)",
      /** #005b66 */
      "35": "var(--leo-color-primitive-teal-35)",
      /** #006875 */
      "40": "var(--leo-color-primitive-teal-40)",
      /** #008393 */
      "50": "var(--leo-color-primitive-teal-50)",
      /** #009fb2 */
      "60": "var(--leo-color-primitive-teal-60)",
      /** #28bcd0 */
      "70": "var(--leo-color-primitive-teal-70)",
      /** #52d7ec */
      "80": "var(--leo-color-primitive-teal-80)",
      /** #9bf0ff */
      "90": "var(--leo-color-primitive-teal-90)",
      /** #d1f7ff */
      "95": "var(--leo-color-primitive-teal-95)",
      /** #edfcff */
      "98": "var(--leo-color-primitive-teal-98)",
      /** #f6fdff */
      "99": "var(--leo-color-primitive-teal-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-teal-100)",
      "toString": [
        ""
      ]
    },
    "blue": {
      /** #000000 */
      "0": "var(--leo-color-primitive-blue-0)",
      /** #001225 */
      "5": "var(--leo-color-primitive-blue-5)",
      /** #001c37 */
      "10": "var(--leo-color-primitive-blue-10)",
      /** #002748 */
      "15": "var(--leo-color-primitive-blue-15)",
      /** #00325a */
      "20": "var(--leo-color-primitive-blue-20)",
      /** #003d6c */
      "25": "var(--leo-color-primitive-blue-25)",
      /** #00497f */
      "30": "var(--leo-color-primitive-blue-30)",
      /** #005492 */
      "35": "var(--leo-color-primitive-blue-35)",
      /** #0061a6 */
      "40": "var(--leo-color-primitive-blue-40)",
      /** #007ad0 */
      "50": "var(--leo-color-primitive-blue-50)",
      /** #2794f3 */
      "60": "var(--leo-color-primitive-blue-60)",
      /** #67afff */
      "70": "var(--leo-color-primitive-blue-70)",
      /** #a0c9ff */
      "80": "var(--leo-color-primitive-blue-80)",
      /** #d2e4ff */
      "90": "var(--leo-color-primitive-blue-90)",
      /** #eaf1ff */
      "95": "var(--leo-color-primitive-blue-95)",
      /** #f8f9ff */
      "98": "var(--leo-color-primitive-blue-98)",
      /** #fdfcff */
      "99": "var(--leo-color-primitive-blue-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-blue-100)",
      "toString": [
        ""
      ]
    },
    "blurple": {
      /** #000000 */
      "0": "var(--leo-color-primitive-blurple-0)",
      /** #05004c */
      "5": "var(--leo-color-primitive-blurple-5)",
      /** #0f0663 */
      "10": "var(--leo-color-primitive-blurple-10)",
      /** #190b7d */
      "15": "var(--leo-color-primitive-blurple-15)",
      /** #251491 */
      "20": "var(--leo-color-primitive-blurple-20)",
      /** #2e19b0 */
      "25": "var(--leo-color-primitive-blurple-25)",
      /** #372cbf */
      "30": "var(--leo-color-primitive-blurple-30)",
      /** #3e37d4 */
      "35": "var(--leo-color-primitive-blurple-35)",
      /** #4a46e0 */
      "40": "var(--leo-color-primitive-blurple-40)",
      /** #6261ff */
      "50": "var(--leo-color-primitive-blurple-50)",
      /** #8283ff */
      "60": "var(--leo-color-primitive-blurple-60)",
      /** #a1a2ff */
      "70": "var(--leo-color-primitive-blurple-70)",
      /** #c1c4ff */
      "80": "var(--leo-color-primitive-blurple-80)",
      /** #dfe1ff */
      "90": "var(--leo-color-primitive-blurple-90)",
      /** #efefff */
      "95": "var(--leo-color-primitive-blurple-95)",
      /** #f8f8ff */
      "98": "var(--leo-color-primitive-blurple-98)",
      /** #fbfbff */
      "99": "var(--leo-color-primitive-blurple-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-blurple-100)",
      "toString": [
        ""
      ]
    },
    "purple": {
      /** #000000 */
      "0": "var(--leo-color-primitive-purple-0)",
      /** #0c0048 */
      "5": "var(--leo-color-primitive-purple-5)",
      /** #160066 */
      "10": "var(--leo-color-primitive-purple-10)",
      /** #1e0083 */
      "15": "var(--leo-color-primitive-purple-15)",
      /** #291391 */
      "20": "var(--leo-color-primitive-purple-20)",
      /** #35249c */
      "25": "var(--leo-color-primitive-purple-25)",
      /** #4132a7 */
      "30": "var(--leo-color-primitive-purple-30)",
      /** #4d40b4 */
      "35": "var(--leo-color-primitive-purple-35)",
      /** #594dc1 */
      "40": "var(--leo-color-primitive-purple-40)",
      /** #7267dc */
      "50": "var(--leo-color-primitive-purple-50)",
      /** #8c81f8 */
      "60": "var(--leo-color-primitive-purple-60)",
      /** #a9a0ff */
      "70": "var(--leo-color-primitive-purple-70)",
      /** #c6bfff */
      "80": "var(--leo-color-primitive-purple-80)",
      /** #e4dfff */
      "90": "var(--leo-color-primitive-purple-90)",
      /** #f3eeff */
      "95": "var(--leo-color-primitive-purple-95)",
      /** #fcf8ff */
      "98": "var(--leo-color-primitive-purple-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-purple-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-purple-100)",
      "toString": [
        ""
      ]
    },
    "pink": {
      /** #000000 */
      "0": "var(--leo-color-primitive-pink-0)",
      /** #2a0015 */
      "5": "var(--leo-color-primitive-pink-5)",
      /** #3e0021 */
      "10": "var(--leo-color-primitive-pink-10)",
      /** #50002d */
      "15": "var(--leo-color-primitive-pink-15)",
      /** #640039 */
      "20": "var(--leo-color-primitive-pink-20)",
      /** #780045 */
      "25": "var(--leo-color-primitive-pink-25)",
      /** #8d0052 */
      "30": "var(--leo-color-primitive-pink-30)",
      /** #a2005f */
      "35": "var(--leo-color-primitive-pink-35)",
      /** #b3186c */
      "40": "var(--leo-color-primitive-pink-40)",
      /** #d43785 */
      "50": "var(--leo-color-primitive-pink-50)",
      /** #f553a0 */
      "60": "var(--leo-color-primitive-pink-60)",
      /** #ff82b6 */
      "70": "var(--leo-color-primitive-pink-70)",
      /** #e6bbcf */
      "80": "var(--leo-color-primitive-pink-80)",
      /** #ffd9e4 */
      "90": "var(--leo-color-primitive-pink-90)",
      /** #ffecf0 */
      "95": "var(--leo-color-primitive-pink-95)",
      /** #fff8f8 */
      "98": "var(--leo-color-primitive-pink-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-pink-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-pink-100)",
      "toString": [
        ""
      ]
    },
    "privateWindow": {
      /** #000000 */
      "0": "var(--leo-color-primitive-private-window-0)",
      /** #160041 */
      "5": "var(--leo-color-primitive-private-window-5)",
      /** #22005d */
      "10": "var(--leo-color-primitive-private-window-10)",
      /** #2e0077 */
      "15": "var(--leo-color-primitive-private-window-15)",
      /** #3a0092 */
      "20": "var(--leo-color-primitive-private-window-20)",
      /** #46199d */
      "25": "var(--leo-color-primitive-private-window-25)",
      /** #512aa9 */
      "30": "var(--leo-color-primitive-private-window-30)",
      /** #5d38b5 */
      "35": "var(--leo-color-primitive-private-window-35)",
      /** #6a45c2 */
      "40": "var(--leo-color-primitive-private-window-40)",
      /** #8360dd */
      "50": "var(--leo-color-primitive-private-window-50)",
      /** #9e7bf9 */
      "60": "var(--leo-color-primitive-private-window-60)",
      /** #b69bff */
      "70": "var(--leo-color-primitive-private-window-70)",
      /** #cfbcff */
      "80": "var(--leo-color-primitive-private-window-80)",
      /** #e9ddff */
      "90": "var(--leo-color-primitive-private-window-90)",
      /** #f6eeff */
      "95": "var(--leo-color-primitive-private-window-95)",
      /** #fdf7ff */
      "98": "var(--leo-color-primitive-private-window-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-private-window-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-private-window-100)",
      "toString": [
        ""
      ]
    },
    "torWindow": {
      /** #000000 */
      "0": "var(--leo-color-primitive-tor-window-0)",
      /** #210033 */
      "5": "var(--leo-color-primitive-tor-window-5)",
      /** #310049 */
      "10": "var(--leo-color-primitive-tor-window-10)",
      /** #40005e */
      "15": "var(--leo-color-primitive-tor-window-15)",
      /** #500074 */
      "20": "var(--leo-color-primitive-tor-window-20)",
      /** #61008b */
      "25": "var(--leo-color-primitive-tor-window-25)",
      /** #7107a1 */
      "30": "var(--leo-color-primitive-tor-window-30)",
      /** #7e20ad */
      "35": "var(--leo-color-primitive-tor-window-35)",
      /** #8c30bb */
      "40": "var(--leo-color-primitive-tor-window-40)",
      /** #a74dd6 */
      "50": "var(--leo-color-primitive-tor-window-50)",
      /** #c369f2 */
      "60": "var(--leo-color-primitive-tor-window-60)",
      /** #d98cff */
      "70": "var(--leo-color-primitive-tor-window-70)",
      /** #e8b3ff */
      "80": "var(--leo-color-primitive-tor-window-80)",
      /** #f6d9ff */
      "90": "var(--leo-color-primitive-tor-window-90)",
      /** #fdebff */
      "95": "var(--leo-color-primitive-tor-window-95)",
      /** #fff7fc */
      "98": "var(--leo-color-primitive-tor-window-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-tor-window-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-tor-window-100)",
      "toString": [
        ""
      ]
    },
    "neutralVariant": {
      /** #000000 */
      "0": "var(--leo-color-primitive-neutral-variant-0)",
      /** #101018 */
      "5": "var(--leo-color-primitive-neutral-variant-5)",
      /** #1b1b23 */
      "10": "var(--leo-color-primitive-neutral-variant-10)",
      /** #25252d */
      "15": "var(--leo-color-primitive-neutral-variant-15)",
      /** #303038 */
      "20": "var(--leo-color-primitive-neutral-variant-20)",
      /** #3b3a43 */
      "25": "var(--leo-color-primitive-neutral-variant-25)",
      /** #47464f */
      "30": "var(--leo-color-primitive-neutral-variant-30)",
      /** #52515b */
      "35": "var(--leo-color-primitive-neutral-variant-35)",
      /** #5e5d67 */
      "40": "var(--leo-color-primitive-neutral-variant-40)",
      /** #777680 */
      "50": "var(--leo-color-primitive-neutral-variant-50)",
      /** #918f9a */
      "60": "var(--leo-color-primitive-neutral-variant-60)",
      /** #acaab4 */
      "70": "var(--leo-color-primitive-neutral-variant-70)",
      /** #c8c5d0 */
      "80": "var(--leo-color-primitive-neutral-variant-80)",
      /** #e4e1ec */
      "90": "var(--leo-color-primitive-neutral-variant-90)",
      /** #f2effa */
      "95": "var(--leo-color-primitive-neutral-variant-95)",
      /** #fcf8ff */
      "98": "var(--leo-color-primitive-neutral-variant-98)",
      /** #fffbff */
      "99": "var(--leo-color-primitive-neutral-variant-99)",
      /** #ffffff */
      "100": "var(--leo-color-primitive-neutral-variant-100)",
      "toString": [
        ""
      ]
    },
    "brands": {
      "toString": [
        ""
      ],
      /** #ff4724 */
      "bat1": "var(--leo-color-primitive-brands-bat-1)",
      /** #9e1f63 */
      "bat2": "var(--leo-color-primitive-brands-bat-2)",
      /** #662d91 */
      "bat3": "var(--leo-color-primitive-brands-bat-3)",
      /** #ff5601 */
      "rorange1": "var(--leo-color-primitive-brands-rorange-1)",
      /** #ff4000 */
      "rorange2": "var(--leo-color-primitive-brands-rorange-2)",
      /** #ff1f01 */
      "rorange3": "var(--leo-color-primitive-brands-rorange-3)",
      /** #4641ee */
      "primaryFixed": "var(--leo-color-primitive-brands-primary-fixed)"
    }
  },
  /** #ffffff */
  "white": "var(--leo-color-white)",
  /** #000000 */
  "black": "var(--leo-color-black)",
  "text": {
    "toString": [
      ""
    ],
    /** #000000 */
    "primary": "var(--leo-color-text-primary)",
    /** #000000 */
    "secondary": "var(--leo-color-text-secondary)",
    /** #000000 */
    "tertiary": "var(--leo-color-text-tertiary)",
    /** #000000 */
    "interactive": "var(--leo-color-text-interactive)",
    /** #000000 */
    "interactiveVisited": "var(--leo-color-text-interactive-visited)",
    /** #21272a80 */
    "disabled": "var(--leo-color-text-disabled)"
  },
  "icon": {
    "toString": [
      ""
    ],
    /** #000000 */
    "default": "var(--leo-color-icon-default)",
    /** #000000 */
    "interactive": "var(--leo-color-icon-interactive)",
    /** #46464a80 */
    "disabled": "var(--leo-color-icon-disabled)"
  },
  "divider": {
    "toString": [
      ""
    ],
    /** #000000 */
    "subtle": "var(--leo-color-divider-subtle)",
    /** #000000 */
    "strong": "var(--leo-color-divider-strong)",
    /** #000000 */
    "interactive": "var(--leo-color-divider-interactive)"
  },
  "container": {
    "toString": [
      ""
    ],
    /** #000000 */
    "background": "var(--leo-color-container-background)",
    /** #000000 */
    "highlight": "var(--leo-color-container-highlight)",
    /** #000000 */
    "interactive": "var(--leo-color-container-interactive)",
    /** #000000 */
    "disabled": "var(--leo-color-container-disabled)"
  },
  "page": {
    "toString": [
      ""
    ],
    /** #000000 */
    "background": "var(--leo-color-page-background)"
  },
  "button": {
    "toString": [
      ""
    ],
    /** #000000 */
    "background": "var(--leo-color-button-background)",
    /** #46464a33 */
    "disabled": "var(--leo-color-button-disabled)",
    /** #000000 */
    "errorBackground": "var(--leo-color-button-error-background)",
    /** #000000 */
    "successBackground": "var(--leo-color-button-success-background)"
  },
  "dialogs": {
    "toString": [
      ""
    ],
    /** #0d121459 */
    "scrimBackground": "var(--leo-color-dialogs-scrim-background)",
    /** #21272a1a */
    "divider": "var(--leo-color-dialogs-divider)",
    /** #0d121459 */
    "buttonTransparentDefault": "var(--leo-color-dialogs-button-transparent-default)",
    /** #0d121480 */
    "buttonTransparentHover": "var(--leo-color-dialogs-button-transparent-hover)"
  },
  "systemfeedback": {
    "toString": [
      ""
    ],
    /** #000000 */
    "infoText": "var(--leo-color-systemfeedback-info-text)",
    /** #000000 */
    "infoIcon": "var(--leo-color-systemfeedback-info-icon)",
    /** #000000 */
    "infoBackground": "var(--leo-color-systemfeedback-info-background)",
    /** #000000 */
    "successText": "var(--leo-color-systemfeedback-success-text)",
    /** #000000 */
    "successIcon": "var(--leo-color-systemfeedback-success-icon)",
    /** #000000 */
    "successBackground": "var(--leo-color-systemfeedback-success-background)",
    /** #000000 */
    "warningText": "var(--leo-color-systemfeedback-warning-text)",
    /** #000000 */
    "warningIcon": "var(--leo-color-systemfeedback-warning-icon)",
    /** #000000 */
    "warningBackground": "var(--leo-color-systemfeedback-warning-background)",
    /** #000000 */
    "errorText": "var(--leo-color-systemfeedback-error-text)",
    /** #000000 */
    "errorIcon": "var(--leo-color-systemfeedback-error-icon)",
    /** #000000 */
    "errorBackground": "var(--leo-color-systemfeedback-error-background)",
    /** #000000 */
    "focusDefault": "var(--leo-color-systemfeedback-focus-default)",
    /** #000000 */
    "focusButtonSpecial": "var(--leo-color-systemfeedback-focus-button-special)"
  },
  "elevation": {
    "toString": [
      ""
    ],
    /** #0000000d */
    "primary": "var(--leo-color-elevation-primary)",
    /** #0000001a */
    "secondary": "var(--leo-color-elevation-secondary)"
  },
  "material": {
    "toString": [
      ""
    ],
    /** #ffffffe8 */
    "thick": "var(--leo-color-material-thick)",
    /** #ffffffba */
    "regular": "var(--leo-color-material-regular)",
    /** #ffffff8c */
    "thin": "var(--leo-color-material-thin)",
    /** #ffffff52 */
    "ultrathin": "var(--leo-color-material-ultrathin)",
    /** #00000014 */
    "separator": "var(--leo-color-material-separator)",
    /** #11111140 */
    "divider": "var(--leo-color-material-divider)",
    /** #000000px */
    "blurThick": "var(--leo-color-material-blur-thick)",
    /** #000000px */
    "blurRegular": "var(--leo-color-material-blur-regular)",
    /** #000000px */
    "blurThin": "var(--leo-color-material-blur-thin)",
    /** #000000px */
    "blurUltrathin": "var(--leo-color-material-blur-ultrathin)"
  },
  /** #000000 */
  "foreground": "var(--leo-color-foreground)",
  "neutral": {
    /** #000000 */
    "10": "var(--leo-color-neutral-10)",
    /** #000000 */
    "20": "var(--leo-color-neutral-20)",
    /** #000000 */
    "30": "var(--leo-color-neutral-30)",
    /** #000000 */
    "40": "var(--leo-color-neutral-40)",
    /** #000000 */
    "50": "var(--leo-color-neutral-50)",
    /** #000000 */
    "60": "var(--leo-color-neutral-60)",
    /** #000000 */
    "70": "var(--leo-color-neutral-70)",
    "toString": [
      ""
    ]
  },
  "primary": {
    /** #000000 */
    "10": "var(--leo-color-primary-10)",
    /** #000000 */
    "20": "var(--leo-color-primary-20)",
    /** #000000 */
    "30": "var(--leo-color-primary-30)",
    /** #000000 */
    "40": "var(--leo-color-primary-40)",
    /** #000000 */
    "50": "var(--leo-color-primary-50)",
    /** #000000 */
    "60": "var(--leo-color-primary-60)",
    /** #000000 */
    "70": "var(--leo-color-primary-70)",
    "toString": [
      ""
    ]
  },
  "secondary": {
    /** #000000 */
    "10": "var(--leo-color-secondary-10)",
    /** #000000 */
    "20": "var(--leo-color-secondary-20)",
    /** #000000 */
    "30": "var(--leo-color-secondary-30)",
    /** #000000 */
    "40": "var(--leo-color-secondary-40)",
    /** #000000 */
    "50": "var(--leo-color-secondary-50)",
    /** #000000 */
    "60": "var(--leo-color-secondary-60)",
    /** #000000 */
    "70": "var(--leo-color-secondary-70)",
    "toString": [
      ""
    ]
  },
  "tertiary": {
    /** #000000 */
    "10": "var(--leo-color-tertiary-10)",
    /** #000000 */
    "20": "var(--leo-color-tertiary-20)",
    /** #000000 */
    "30": "var(--leo-color-tertiary-30)",
    /** #000000 */
    "40": "var(--leo-color-tertiary-40)",
    /** #000000 */
    "50": "var(--leo-color-tertiary-50)",
    /** #000000 */
    "60": "var(--leo-color-tertiary-60)",
    /** #000000 */
    "70": "var(--leo-color-tertiary-70)",
    "toString": [
      ""
    ]
  },
  "red": {
    /** #000000 */
    "10": "var(--leo-color-red-10)",
    /** #000000 */
    "20": "var(--leo-color-red-20)",
    /** #000000 */
    "30": "var(--leo-color-red-30)",
    /** #000000 */
    "40": "var(--leo-color-red-40)",
    /** #000000 */
    "50": "var(--leo-color-red-50)",
    /** #000000 */
    "60": "var(--leo-color-red-60)",
    /** #000000 */
    "70": "var(--leo-color-red-70)",
    "toString": [
      ""
    ]
  },
  "orange": {
    /** #000000 */
    "10": "var(--leo-color-orange-10)",
    /** #000000 */
    "20": "var(--leo-color-orange-20)",
    /** #000000 */
    "30": "var(--leo-color-orange-30)",
    /** #000000 */
    "40": "var(--leo-color-orange-40)",
    /** #000000 */
    "50": "var(--leo-color-orange-50)",
    /** #000000 */
    "60": "var(--leo-color-orange-60)",
    /** #000000 */
    "70": "var(--leo-color-orange-70)",
    "toString": [
      ""
    ]
  },
  "yellow": {
    /** #000000 */
    "10": "var(--leo-color-yellow-10)",
    /** #000000 */
    "20": "var(--leo-color-yellow-20)",
    /** #000000 */
    "30": "var(--leo-color-yellow-30)",
    /** #000000 */
    "40": "var(--leo-color-yellow-40)",
    /** #000000 */
    "50": "var(--leo-color-yellow-50)",
    /** #000000 */
    "60": "var(--leo-color-yellow-60)",
    /** #000000 */
    "70": "var(--leo-color-yellow-70)",
    "toString": [
      ""
    ]
  },
  "green": {
    /** #000000 */
    "10": "var(--leo-color-green-10)",
    /** #000000 */
    "20": "var(--leo-color-green-20)",
    /** #000000 */
    "30": "var(--leo-color-green-30)",
    /** #000000 */
    "40": "var(--leo-color-green-40)",
    /** #000000 */
    "50": "var(--leo-color-green-50)",
    /** #000000 */
    "60": "var(--leo-color-green-60)",
    /** #000000 */
    "70": "var(--leo-color-green-70)",
    "toString": [
      ""
    ]
  },
  "teal": {
    /** #000000 */
    "10": "var(--leo-color-teal-10)",
    /** #000000 */
    "20": "var(--leo-color-teal-20)",
    /** #000000 */
    "30": "var(--leo-color-teal-30)",
    /** #000000 */
    "40": "var(--leo-color-teal-40)",
    /** #000000 */
    "50": "var(--leo-color-teal-50)",
    /** #000000 */
    "60": "var(--leo-color-teal-60)",
    /** #000000 */
    "70": "var(--leo-color-teal-70)",
    "toString": [
      ""
    ]
  },
  "blue": {
    /** #000000 */
    "10": "var(--leo-color-blue-10)",
    /** #000000 */
    "20": "var(--leo-color-blue-20)",
    /** #000000 */
    "30": "var(--leo-color-blue-30)",
    /** #000000 */
    "40": "var(--leo-color-blue-40)",
    /** #000000 */
    "50": "var(--leo-color-blue-50)",
    /** #000000 */
    "60": "var(--leo-color-blue-60)",
    /** #000000 */
    "70": "var(--leo-color-blue-70)",
    "toString": [
      ""
    ]
  },
  "blurple": {
    /** #000000 */
    "10": "var(--leo-color-blurple-10)",
    /** #000000 */
    "20": "var(--leo-color-blurple-20)",
    /** #000000 */
    "30": "var(--leo-color-blurple-30)",
    /** #000000 */
    "40": "var(--leo-color-blurple-40)",
    /** #000000 */
    "50": "var(--leo-color-blurple-50)",
    /** #000000 */
    "60": "var(--leo-color-blurple-60)",
    /** #000000 */
    "70": "var(--leo-color-blurple-70)",
    "toString": [
      ""
    ]
  },
  "purple": {
    /** #000000 */
    "10": "var(--leo-color-purple-10)",
    /** #000000 */
    "20": "var(--leo-color-purple-20)",
    /** #000000 */
    "30": "var(--leo-color-purple-30)",
    /** #000000 */
    "40": "var(--leo-color-purple-40)",
    /** #000000 */
    "50": "var(--leo-color-purple-50)",
    /** #000000 */
    "60": "var(--leo-color-purple-60)",
    /** #000000 */
    "70": "var(--leo-color-purple-70)",
    "toString": [
      ""
    ]
  },
  "pink": {
    /** #000000 */
    "10": "var(--leo-color-pink-10)",
    /** #000000 */
    "20": "var(--leo-color-pink-20)",
    /** #000000 */
    "30": "var(--leo-color-pink-30)",
    /** #000000 */
    "40": "var(--leo-color-pink-40)",
    /** #000000 */
    "50": "var(--leo-color-pink-50)",
    /** #000000 */
    "60": "var(--leo-color-pink-60)",
    /** #000000 */
    "70": "var(--leo-color-pink-70)",
    "toString": [
      ""
    ]
  },
  "privateWindow": {
    /** #000000 */
    "10": "var(--leo-color-private-window-10)",
    /** #000000 */
    "20": "var(--leo-color-private-window-20)",
    /** #000000 */
    "30": "var(--leo-color-private-window-30)",
    /** #000000 */
    "40": "var(--leo-color-private-window-40)",
    /** #000000 */
    "50": "var(--leo-color-private-window-50)",
    /** #000000 */
    "60": "var(--leo-color-private-window-60)",
    /** #000000 */
    "70": "var(--leo-color-private-window-70)",
    "toString": [
      ""
    ]
  },
  "torWindow": {
    /** #000000 */
    "10": "var(--leo-color-tor-window-10)",
    /** #000000 */
    "20": "var(--leo-color-tor-window-20)",
    /** #000000 */
    "30": "var(--leo-color-tor-window-30)",
    /** #000000 */
    "40": "var(--leo-color-tor-window-40)",
    /** #000000 */
    "50": "var(--leo-color-tor-window-50)",
    /** #000000 */
    "60": "var(--leo-color-tor-window-60)",
    /** #000000 */
    "70": "var(--leo-color-tor-window-70)",
    "toString": [
      ""
    ]
  },
  "neutralVariant": {
    /** #000000 */
    "10": "var(--leo-color-neutral-variant-10)",
    /** #000000 */
    "20": "var(--leo-color-neutral-variant-20)",
    /** #000000 */
    "30": "var(--leo-color-neutral-variant-30)",
    /** #000000 */
    "40": "var(--leo-color-neutral-variant-40)",
    /** #000000 */
    "50": "var(--leo-color-neutral-variant-50)",
    /** #000000 */
    "60": "var(--leo-color-neutral-variant-60)",
    /** #000000 */
    "70": "var(--leo-color-neutral-variant-70)",
    "toString": [
      ""
    ]
  },
  "schemes": {
    "toString": [
      ""
    ],
    /** #000000 */
    "surfaceTint": "var(--leo-color-schemes-surface-tint)",
    /** #000000 */
    "primary": "var(--leo-color-schemes-primary)",
    /** #000000 */
    "onPrimary": "var(--leo-color-schemes-on-primary)",
    /** #000000 */
    "primaryContainer": "var(--leo-color-schemes-primary-container)",
    /** #000000 */
    "onPrimaryContainer": "var(--leo-color-schemes-on-primary-container)",
    /** #000000 */
    "secondary": "var(--leo-color-schemes-secondary)",
    /** #000000 */
    "onSecondary": "var(--leo-color-schemes-on-secondary)",
    /** #000000 */
    "secondaryContainer": "var(--leo-color-schemes-secondary-container)",
    /** #000000 */
    "onSecondaryContainer": "var(--leo-color-schemes-on-secondary-container)",
    /** #000000 */
    "tertiary": "var(--leo-color-schemes-tertiary)",
    /** #000000 */
    "onTertiary": "var(--leo-color-schemes-on-tertiary)",
    /** #000000 */
    "tertiaryContainer": "var(--leo-color-schemes-tertiary-container)",
    /** #000000 */
    "onTertiaryContainer": "var(--leo-color-schemes-on-tertiary-container)",
    /** #000000 */
    "error": "var(--leo-color-schemes-error)",
    /** #000000 */
    "onError": "var(--leo-color-schemes-on-error)",
    /** #000000 */
    "errorContainer": "var(--leo-color-schemes-error-container)",
    /** #000000 */
    "onErrorContainer": "var(--leo-color-schemes-on-error-container)",
    /** #000000 */
    "background": "var(--leo-color-schemes-background)",
    /** #000000 */
    "onBackground": "var(--leo-color-schemes-on-background)",
    /** #000000 */
    "surface": "var(--leo-color-schemes-surface)",
    /** #000000 */
    "onSurface": "var(--leo-color-schemes-on-surface)",
    /** #000000 */
    "surfaceVariant": "var(--leo-color-schemes-surface-variant)",
    /** #000000 */
    "onSurfaceVariant": "var(--leo-color-schemes-on-surface-variant)",
    /** #000000 */
    "outline": "var(--leo-color-schemes-outline)",
    /** #000000 */
    "outlineVariant": "var(--leo-color-schemes-outline-variant)",
    /** #000000 */
    "shadow": "var(--leo-color-schemes-shadow)",
    /** #000000 */
    "scrim": "var(--leo-color-schemes-scrim)",
    /** #000000 */
    "inverseSurface": "var(--leo-color-schemes-inverse-surface)",
    /** #000000 */
    "inverseOnSurface": "var(--leo-color-schemes-inverse-on-surface)",
    /** #000000 */
    "inversePrimary": "var(--leo-color-schemes-inverse-primary)",
    /** #000000 */
    "primaryFixed": "var(--leo-color-schemes-primary-fixed)",
    /** #000000 */
    "onPrimaryFixed": "var(--leo-color-schemes-on-primary-fixed)",
    /** #000000 */
    "primaryFixedDim": "var(--leo-color-schemes-primary-fixed-dim)",
    /** #000000 */
    "onPrimaryFixedVariant": "var(--leo-color-schemes-on-primary-fixed-variant)",
    /** #000000 */
    "secondaryFixed": "var(--leo-color-schemes-secondary-fixed)",
    /** #000000 */
    "onSecondaryFixed": "var(--leo-color-schemes-on-secondary-fixed)",
    /** #000000 */
    "secondaryFixedDim": "var(--leo-color-schemes-secondary-fixed-dim)",
    /** #000000 */
    "onSecondaryFixedVariant": "var(--leo-color-schemes-on-secondary-fixed-variant)",
    /** #000000 */
    "tertiaryFixed": "var(--leo-color-schemes-tertiary-fixed)",
    /** #000000 */
    "onTertiaryFixed": "var(--leo-color-schemes-on-tertiary-fixed)",
    /** #000000 */
    "tertiaryFixedDim": "var(--leo-color-schemes-tertiary-fixed-dim)",
    /** #000000 */
    "onTertiaryFixedVariant": "var(--leo-color-schemes-on-tertiary-fixed-variant)",
    /** #000000 */
    "surfaceDim": "var(--leo-color-schemes-surface-dim)",
    /** #000000 */
    "surfaceBright": "var(--leo-color-schemes-surface-bright)",
    /** #000000 */
    "surfaceContainerLowest": "var(--leo-color-schemes-surface-container-lowest)",
    /** #000000 */
    "surfaceContainerLow": "var(--leo-color-schemes-surface-container-low)",
    /** #000000 */
    "surfaceContainer": "var(--leo-color-schemes-surface-container)",
    /** #000000 */
    "surfaceContainerHigh": "var(--leo-color-schemes-surface-container-high)",
    /** #000000 */
    "surfaceContainerHighest": "var(--leo-color-schemes-surface-container-highest)"
  },
  "ios": {
    "toString": [
      ""
    ],
    "browser": {
      "toString": [
        ""
      ],
      /** #000000 */
      "chromeBackgroundIos": "var(--leo-color-ios-browser-chrome-background-ios)",
      /** #000000 */
      "omnibarBackground": "var(--leo-color-ios-browser-omnibar-background)",
      /** #000000 */
      "backgroundIos": "var(--leo-color-ios-browser-background-ios)",
      /** #000000 */
      "elevatedIos": "var(--leo-color-ios-browser-elevated-ios)",
      /** #7676801f */
      "containerHighlightIos": "var(--leo-color-ios-browser-container-highlight-ios)"
    }
  },
  "desktopbrowser": {
    "toString": [
      ""
    ],
    /** #000000 */
    "chromeBackgroundDesktop": "var(--leo-color-desktopbrowser-chrome-background-desktop)",
    /** #000000 */
    "chromeMenuBackground": "var(--leo-color-desktopbrowser-chrome-menu-background)",
    "tabbar": {
      "toString": [
        ""
      ],
      /** #000000 */
      "background": "var(--leo-color-desktopbrowser-tabbar-background)",
      /** #000000 */
      "pinnedTabOutlineHorizontal": "var(--leo-color-desktopbrowser-tabbar-pinned-tab-outline-horizontal)",
      /** #000000 */
      "pinnedTabOutlineVertical": "var(--leo-color-desktopbrowser-tabbar-pinned-tab-outline-vertical)",
      /** #000000 */
      "hoverTabBackground": "var(--leo-color-desktopbrowser-tabbar-hover-tab-background)",
      /** #000000 */
      "activeButtonBackground": "var(--leo-color-desktopbrowser-tabbar-active-button-background)",
      /** #000000 */
      "activeTabVertical": "var(--leo-color-desktopbrowser-tabbar-active-tab-vertical)",
      /** #000000 */
      "activeTabHorizontal": "var(--leo-color-desktopbrowser-tabbar-active-tab-horizontal)",
      /** #000000 */
      "splitViewHorizontal": "var(--leo-color-desktopbrowser-tabbar-split-view-horizontal)",
      /** #000000 */
      "splitViewVertical": "var(--leo-color-desktopbrowser-tabbar-split-view-vertical)"
    },
    "omnibar": {
      "toString": [
        ""
      ],
      /** #000000 */
      "backgroundDesktop": "var(--leo-color-desktopbrowser-omnibar-background-desktop)",
      /** #000000 */
      "backgroundActive": "var(--leo-color-desktopbrowser-omnibar-background-active)",
      /** #000000 */
      "backgroundSelectedItem": "var(--leo-color-desktopbrowser-omnibar-background-selected-item)",
      /** #000000 */
      "splitViewIconBackground": "var(--leo-color-desktopbrowser-omnibar-split-view-icon-background)",
      /** #000000 */
      "splitViewIconActive": "var(--leo-color-desktopbrowser-omnibar-split-view-icon-active)",
      "button": {
        "toString": [
          ""
        ],
        /** #000000 */
        "hover": "var(--leo-color-desktopbrowser-omnibar-button-hover)",
        /** #000000 */
        "active": "var(--leo-color-desktopbrowser-omnibar-button-active)"
      },
      /** #000000 */
      "divider": "var(--leo-color-desktopbrowser-omnibar-divider)"
    },
    "toolbar": {
      "toString": [
        ""
      ],
      "button": {
        "toString": [
          ""
        ],
        /** #000000 */
        "hover": "var(--leo-color-desktopbrowser-toolbar-button-hover)",
        /** #000000 */
        "active": "var(--leo-color-desktopbrowser-toolbar-button-active)",
        /** #000000 */
        "outline": "var(--leo-color-desktopbrowser-toolbar-button-outline)"
      }
    }
  }
} as const 
export const gradient = {
  "toString": [
    ""
  ],
  /** linear-gradient(179.24640957924876deg, rgb(255, 64, 0) 0%, rgb(255, 31, 1) 100%) */
  "hero": "var(--leo-gradient-hero)",
  /** linear-gradient(321.5013256215316deg, rgb(250, 114, 80) 3%, rgb(255, 24, 147) 40%, rgb(167, 138, 255) 99%) */
  "iconsActive": "var(--leo-gradient-icons-active)",
  /** linear-gradient(134.9999995731132deg, rgb(70, 65, 238) 0%, rgb(98, 97, 255) 100%) */
  "panelBackground": "var(--leo-gradient-panel-background)",
  /** linear-gradient(134.9999995731132deg, rgb(70, 65, 238) 0%, rgb(114, 103, 220) 100%) */
  "panelBackgroundAlternative": "var(--leo-gradient-panel-background-alternative)",
  /** linear-gradient(127.06278645486384deg, rgb(0, 122, 208) 0%, rgb(39, 148, 243) 100%) */
  "premiumindicator": "var(--leo-gradient-premiumindicator)",
  /** linear-gradient(314.9161155270921deg, rgb(236, 19, 73) 0%, rgb(212, 17, 115) 100%) */
  "toolbarBackground": "var(--leo-gradient-toolbar-background)",
  /** linear-gradient(90deg, rgb(255, 86, 1) 0%, rgb(255, 64, 0) 50%, rgb(255, 31, 1) 100%) */
  "braveRelease": "var(--leo-gradient-brave-release)",
  /** linear-gradient(90deg, rgb(0, 153, 229) 0%, rgb(0, 89, 178) 50%, rgb(0, 43, 128) 100%) */
  "braveBeta": "var(--leo-gradient-brave-beta)",
  /** linear-gradient(90deg, rgb(144, 30, 225) 0%, rgb(101, 20, 158) 51%, rgb(72, 15, 113) 100%) */
  "braveNightly": "var(--leo-gradient-brave-nightly)",
  /** linear-gradient(90deg, rgb(108, 116, 147) 0%, rgb(87, 93, 117) 51%, rgb(54, 58, 73) 100%) */
  "braveDebug": "var(--leo-gradient-brave-debug)",
  /** linear-gradient(138.21219643124405deg, rgb(134, 10, 194) 7.000000000000001%, rgb(230, 25, 135) 68%, rgb(255, 71, 26) 96%) */
  "heroLegacy": "var(--leo-gradient-hero-legacy)"
} as const 
export const font = {
  "toString": [
    ""
  ],
  "components": {
    "toString": [
      ""
    ],
    /** 700 10px/10px Inter Variable */
    "label": "var(--leo-font-components-label)",
    /** 600 13px/16px Inter Variable */
    "tableheader": "var(--leo-font-components-tableheader)",
    /** 600 14px/22px Poppins */
    "navbutton": "var(--leo-font-components-navbutton)",
    /** 600 12px/20px Poppins */
    "buttonSmall": "var(--leo-font-components-button-small)",
    /** 600 13px/20px Poppins */
    "buttonDefault": "var(--leo-font-components-button-default)",
    /** 600 15px/22px Poppins */
    "buttonLarge": "var(--leo-font-components-button-large)",
    /** 600 15px/22px Poppins */
    "buttonJumbo": "var(--leo-font-components-button-jumbo)",
    /** 600 36px/46px Poppins */
    "numbersLarge": "var(--leo-font-components-numbers-large)"
  },
  "heading": {
    "toString": [
      ""
    ],
    /** 600 46px/54px Poppins */
    "display1": "var(--leo-font-heading-display1)",
    /** 600 40px/48px Poppins */
    "display2": "var(--leo-font-heading-display2)",
    /** 600 34px/40px Poppins */
    "display3": "var(--leo-font-heading-display3)",
    /** 600 32px/42px Poppins */
    "h1": "var(--leo-font-heading-h1)",
    /** 600 28px/36px Poppins */
    "h2": "var(--leo-font-heading-h2)",
    /** 600 22px/28px Poppins */
    "h3": "var(--leo-font-heading-h3)",
    /** 600 16px/26px Poppins */
    "h4": "var(--leo-font-heading-h4)"
  },
  "large": {
    "toString": [
      ""
    ],
    /** 400 16px/24px Inter Variable */
    "regular": "var(--leo-font-large-regular)",
    /** 600 16px/24px Inter Variable */
    "semibold": "var(--leo-font-large-semibold)",
    /** 400 16px/24px Inter Variable */
    "link": "var(--leo-font-large-link)"
  },
  "default": {
    "toString": [
      ""
    ],
    /** 400 14px/22px Inter Variable */
    "regular": "var(--leo-font-default-regular)",
    /** 600 14px/22px Inter Variable */
    "semibold": "var(--leo-font-default-semibold)",
    /** 400 14px/22px Inter Variable */
    "link": "var(--leo-font-default-link)"
  },
  "small": {
    "toString": [
      ""
    ],
    /** 400 12px/18px Inter Variable */
    "regular": "var(--leo-font-small-regular)",
    /** 600 12px/18px Inter Variable */
    "semibold": "var(--leo-font-small-semibold)",
    /** 400 12px/18px Inter Variable */
    "link": "var(--leo-font-small-link)"
  },
  "xSmall": {
    "toString": [
      ""
    ],
    /** 400 11px/16px Inter Variable */
    "regular": "var(--leo-font-x-small-regular)",
    /** 600 11px/16px Inter Variable */
    "semibold": "var(--leo-font-x-small-semibold)",
    /** 400 11px/16px Inter Variable */
    "link": "var(--leo-font-x-small-link)"
  },
  "monospace": {
    "toString": [
      ""
    ],
    /** 400 16px/24px DM Mono */
    "large": "var(--leo-font-monospace-large)",
    /** 400 14px/22px DM Mono */
    "default": "var(--leo-font-monospace-default)",
    /** 400 12px/18px DM Mono */
    "small": "var(--leo-font-monospace-small)"
  }
} as const 
export const effect = {
  "toString": [
    ""
  ],
  /** inset 0px 0px 1px 0px rgba(255, 255, 255, 0.1), 0px 1px 1px 0px rgba(0, 0, 0, 0.05), 0px 4px 8px 0px rgba(0, 0, 0, 0.1) */
  "notificationbackdrop": "var(--leo-effect-notificationbackdrop)",
  /** inset 0px 0px 0px 2px rgb(130, 131, 255) */
  "focusState": "var(--leo-effect-focus-state)",
  "elevation": {
    "toString": [
      ""
    ],
    /** 0px 1px 0px 0px rgba(0, 0, 0, 0.05), 0px 1px 4px 0px rgba(0, 0, 0, 0.1) */
    "01": "var(--leo-effect-elevation-01)",
    /** 0px 1px 0px 0px rgba(0, 0, 0, 0.05), 0px 4px 8px -1px rgba(0, 0, 0, 0.1) */
    "02": "var(--leo-effect-elevation-02)",
    /** 0px 1px 0px 0px rgba(0, 0, 0, 0.05), 0px 4px 16px -2px rgba(0, 0, 0, 0.1) */
    "03": "var(--leo-effect-elevation-03)",
    /** 0px 1px 0px 0px rgba(0, 0, 0, 0.05), 0px 4px 24px -1px rgba(0, 0, 0, 0.1) */
    "04": "var(--leo-effect-elevation-04)",
    /** 0px 1px 0px 0px rgba(0, 0, 0, 0.05), 0px 8px 36px -2px rgba(0, 0, 0, 0.1) */
    "05": "var(--leo-effect-elevation-05)"
  },
  "hardShadow": {
    "toString": [
      ""
    ],
    /** 4px 4px 0px 0px rgb(98, 97, 255) */
    "01": "var(--leo-effect-hard-shadow-01)",
    /** 8px 8px 0px 0px rgb(98, 97, 255) */
    "02": "var(--leo-effect-hard-shadow-02)"
  },
  "material": {
    "toString": [
      ""
    ],
    "thickShadow": {
      /** 0px 0px 16px 0px rgba(0, 0, 0, 0.15) */
      "1": "var(--leo-effect-material-thick-shadow-1)",
      "toString": [
        ""
      ]
    },
    "regularShadow": {
      /** 0px 0px 16px 0px rgba(0, 0, 0, 0.15) */
      "1": "var(--leo-effect-material-regular-shadow-1)",
      "toString": [
        ""
      ]
    },
    "thinShadow": {
      /** 0px 0px 16px 0px rgba(0, 0, 0, 0.15) */
      "1": "var(--leo-effect-material-thin-shadow-1)",
      "toString": [
        ""
      ]
    },
    "ultrathinShadow": {
      /** 0px 0px 16px 0px rgba(0, 0, 0, 0.15) */
      "1": "var(--leo-effect-material-ultrathin-shadow-1)",
      "toString": [
        ""
      ]
    }
  }
} as const 
export const typography = {
  "toString": [
    ""
  ],
  "components": {
    "toString": [
      ""
    ],
    "label": {
      "toString": [
        ""
      ],
      /** 10px */
      "fontSize": "var(--leo-typography-components-label-font-size)",
      /** 700px */
      "fontWeight": "var(--leo-typography-components-label-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-components-label-letter-spacing)",
      /** 10px */
      "lineHeight": "var(--leo-typography-components-label-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-label-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-label-paragraph-spacing)"
    },
    "tableheader": {
      "toString": [
        ""
      ],
      /** 13px */
      "fontSize": "var(--leo-typography-components-tableheader-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-tableheader-font-weight)",
      /** -0.1px */
      "letterSpacing": "var(--leo-typography-components-tableheader-letter-spacing)",
      /** 16px */
      "lineHeight": "var(--leo-typography-components-tableheader-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-tableheader-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-tableheader-paragraph-spacing)"
    },
    "navbutton": {
      "toString": [
        ""
      ],
      /** 14px */
      "fontSize": "var(--leo-typography-components-navbutton-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-navbutton-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-components-navbutton-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-components-navbutton-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-navbutton-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-navbutton-paragraph-spacing)"
    },
    "buttonSmall": {
      "toString": [
        ""
      ],
      /** 12px */
      "fontSize": "var(--leo-typography-components-button-small-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-button-small-font-weight)",
      /** 0.1px */
      "letterSpacing": "var(--leo-typography-components-button-small-letter-spacing)",
      /** 20px */
      "lineHeight": "var(--leo-typography-components-button-small-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-button-small-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-button-small-paragraph-spacing)"
    },
    "buttonDefault": {
      "toString": [
        ""
      ],
      /** 13px */
      "fontSize": "var(--leo-typography-components-button-default-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-button-default-font-weight)",
      /** 0.1px */
      "letterSpacing": "var(--leo-typography-components-button-default-letter-spacing)",
      /** 20px */
      "lineHeight": "var(--leo-typography-components-button-default-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-button-default-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-button-default-paragraph-spacing)"
    },
    "buttonLarge": {
      "toString": [
        ""
      ],
      /** 15px */
      "fontSize": "var(--leo-typography-components-button-large-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-button-large-font-weight)",
      /** 0.1px */
      "letterSpacing": "var(--leo-typography-components-button-large-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-components-button-large-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-button-large-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-button-large-paragraph-spacing)"
    },
    "buttonJumbo": {
      "toString": [
        ""
      ],
      /** 15px */
      "fontSize": "var(--leo-typography-components-button-jumbo-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-button-jumbo-font-weight)",
      /** 0.1px */
      "letterSpacing": "var(--leo-typography-components-button-jumbo-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-components-button-jumbo-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-button-jumbo-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-button-jumbo-paragraph-spacing)"
    },
    "numbersLarge": {
      "toString": [
        ""
      ],
      /** 36px */
      "fontSize": "var(--leo-typography-components-numbers-large-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-components-numbers-large-font-weight)",
      /** -0.72px */
      "letterSpacing": "var(--leo-typography-components-numbers-large-letter-spacing)",
      /** 46px */
      "lineHeight": "var(--leo-typography-components-numbers-large-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-components-numbers-large-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-components-numbers-large-paragraph-spacing)"
    }
  },
  "heading": {
    "toString": [
      ""
    ],
    "display1": {
      "toString": [
        ""
      ],
      /** 46px */
      "fontSize": "var(--leo-typography-heading-display1-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-display1-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-display1-letter-spacing)",
      /** 54px */
      "lineHeight": "var(--leo-typography-heading-display1-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-display1-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-display1-paragraph-spacing)"
    },
    "display2": {
      "toString": [
        ""
      ],
      /** 40px */
      "fontSize": "var(--leo-typography-heading-display2-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-display2-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-display2-letter-spacing)",
      /** 48px */
      "lineHeight": "var(--leo-typography-heading-display2-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-display2-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-display2-paragraph-spacing)"
    },
    "display3": {
      "toString": [
        ""
      ],
      /** 34px */
      "fontSize": "var(--leo-typography-heading-display3-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-display3-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-display3-letter-spacing)",
      /** 40px */
      "lineHeight": "var(--leo-typography-heading-display3-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-display3-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-display3-paragraph-spacing)"
    },
    "h1": {
      "toString": [
        ""
      ],
      /** 32px */
      "fontSize": "var(--leo-typography-heading-h1-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-h1-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-h1-letter-spacing)",
      /** 42px */
      "lineHeight": "var(--leo-typography-heading-h1-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-h1-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-h1-paragraph-spacing)"
    },
    "h2": {
      "toString": [
        ""
      ],
      /** 28px */
      "fontSize": "var(--leo-typography-heading-h2-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-h2-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-h2-letter-spacing)",
      /** 36px */
      "lineHeight": "var(--leo-typography-heading-h2-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-h2-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-h2-paragraph-spacing)"
    },
    "h3": {
      "toString": [
        ""
      ],
      /** 22px */
      "fontSize": "var(--leo-typography-heading-h3-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-h3-font-weight)",
      /** -0.5px */
      "letterSpacing": "var(--leo-typography-heading-h3-letter-spacing)",
      /** 28px */
      "lineHeight": "var(--leo-typography-heading-h3-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-h3-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-h3-paragraph-spacing)"
    },
    "h4": {
      "toString": [
        ""
      ],
      /** 16px */
      "fontSize": "var(--leo-typography-heading-h4-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-heading-h4-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-heading-h4-letter-spacing)",
      /** 26px */
      "lineHeight": "var(--leo-typography-heading-h4-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-heading-h4-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-heading-h4-paragraph-spacing)"
    }
  },
  "large": {
    "toString": [
      ""
    ],
    "regular": {
      "toString": [
        ""
      ],
      /** 16px */
      "fontSize": "var(--leo-typography-large-regular-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-large-regular-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-large-regular-letter-spacing)",
      /** 24px */
      "lineHeight": "var(--leo-typography-large-regular-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-large-regular-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-large-regular-paragraph-spacing)"
    },
    "semibold": {
      "toString": [
        ""
      ],
      /** 16px */
      "fontSize": "var(--leo-typography-large-semibold-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-large-semibold-font-weight)",
      /** 0.08px */
      "letterSpacing": "var(--leo-typography-large-semibold-letter-spacing)",
      /** 24px */
      "lineHeight": "var(--leo-typography-large-semibold-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-large-semibold-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-large-semibold-paragraph-spacing)"
    },
    "link": {
      "toString": [
        ""
      ],
      /** 16px */
      "fontSize": "var(--leo-typography-large-link-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-large-link-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-large-link-letter-spacing)",
      /** 24px */
      "lineHeight": "var(--leo-typography-large-link-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-large-link-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-large-link-paragraph-spacing)"
    }
  },
  "default": {
    "toString": [
      ""
    ],
    "regular": {
      "toString": [
        ""
      ],
      /** 14px */
      "fontSize": "var(--leo-typography-default-regular-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-default-regular-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-default-regular-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-default-regular-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-default-regular-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-default-regular-paragraph-spacing)"
    },
    "semibold": {
      "toString": [
        ""
      ],
      /** 14px */
      "fontSize": "var(--leo-typography-default-semibold-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-default-semibold-font-weight)",
      /** 0.07px */
      "letterSpacing": "var(--leo-typography-default-semibold-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-default-semibold-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-default-semibold-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-default-semibold-paragraph-spacing)"
    },
    "link": {
      "toString": [
        ""
      ],
      /** 14px */
      "fontSize": "var(--leo-typography-default-link-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-default-link-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-default-link-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-default-link-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-default-link-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-default-link-paragraph-spacing)"
    }
  },
  "small": {
    "toString": [
      ""
    ],
    "regular": {
      "toString": [
        ""
      ],
      /** 12px */
      "fontSize": "var(--leo-typography-small-regular-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-small-regular-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-small-regular-letter-spacing)",
      /** 18px */
      "lineHeight": "var(--leo-typography-small-regular-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-small-regular-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-small-regular-paragraph-spacing)"
    },
    "semibold": {
      "toString": [
        ""
      ],
      /** 12px */
      "fontSize": "var(--leo-typography-small-semibold-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-small-semibold-font-weight)",
      /** 0.06px */
      "letterSpacing": "var(--leo-typography-small-semibold-letter-spacing)",
      /** 18px */
      "lineHeight": "var(--leo-typography-small-semibold-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-small-semibold-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-small-semibold-paragraph-spacing)"
    },
    "link": {
      "toString": [
        ""
      ],
      /** 12px */
      "fontSize": "var(--leo-typography-small-link-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-small-link-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-small-link-letter-spacing)",
      /** 18px */
      "lineHeight": "var(--leo-typography-small-link-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-small-link-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-small-link-paragraph-spacing)"
    }
  },
  "xSmall": {
    "toString": [
      ""
    ],
    "regular": {
      "toString": [
        ""
      ],
      /** 11px */
      "fontSize": "var(--leo-typography-x-small-regular-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-x-small-regular-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-x-small-regular-letter-spacing)",
      /** 16px */
      "lineHeight": "var(--leo-typography-x-small-regular-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-x-small-regular-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-x-small-regular-paragraph-spacing)"
    },
    "semibold": {
      "toString": [
        ""
      ],
      /** 11px */
      "fontSize": "var(--leo-typography-x-small-semibold-font-size)",
      /** 600px */
      "fontWeight": "var(--leo-typography-x-small-semibold-font-weight)",
      /** 0.055px */
      "letterSpacing": "var(--leo-typography-x-small-semibold-letter-spacing)",
      /** 16px */
      "lineHeight": "var(--leo-typography-x-small-semibold-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-x-small-semibold-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-x-small-semibold-paragraph-spacing)"
    },
    "link": {
      "toString": [
        ""
      ],
      /** 11px */
      "fontSize": "var(--leo-typography-x-small-link-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-x-small-link-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-x-small-link-letter-spacing)",
      /** 16px */
      "lineHeight": "var(--leo-typography-x-small-link-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-x-small-link-paragraph-indent)",
      /** 8px */
      "paragraphSpacing": "var(--leo-typography-x-small-link-paragraph-spacing)"
    }
  },
  "monospace": {
    "toString": [
      ""
    ],
    "large": {
      "toString": [
        ""
      ],
      /** 16px */
      "fontSize": "var(--leo-typography-monospace-large-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-monospace-large-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-monospace-large-letter-spacing)",
      /** 24px */
      "lineHeight": "var(--leo-typography-monospace-large-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-monospace-large-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-monospace-large-paragraph-spacing)"
    },
    "default": {
      "toString": [
        ""
      ],
      /** 14px */
      "fontSize": "var(--leo-typography-monospace-default-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-monospace-default-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-monospace-default-letter-spacing)",
      /** 22px */
      "lineHeight": "var(--leo-typography-monospace-default-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-monospace-default-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-monospace-default-paragraph-spacing)"
    },
    "small": {
      "toString": [
        ""
      ],
      /** 12px */
      "fontSize": "var(--leo-typography-monospace-small-font-size)",
      /** 400px */
      "fontWeight": "var(--leo-typography-monospace-small-font-weight)",
      /** 0 */
      "letterSpacing": "var(--leo-typography-monospace-small-letter-spacing)",
      /** 18px */
      "lineHeight": "var(--leo-typography-monospace-small-line-height)",
      /** 0 */
      "paragraphIndent": "var(--leo-typography-monospace-small-paragraph-indent)",
      /** 0 */
      "paragraphSpacing": "var(--leo-typography-monospace-small-paragraph-spacing)"
    }
  },
  "mobile": {
    "toString": [
      ""
    ],
    "size": {
      "toString": [
        ""
      ],
      /** 16px */
      "large": "var(--leo-typography-mobile-size-large)",
      /** 14px */
      "default": "var(--leo-typography-mobile-size-default)",
      /** 12px */
      "small": "var(--leo-typography-mobile-size-small)",
      /** 11px */
      "xSmall": "var(--leo-typography-mobile-size-x-small)",
      /** 36px */
      "display1": "var(--leo-typography-mobile-size-display1)",
      /** 32px */
      "display2": "var(--leo-typography-mobile-size-display2)",
      /** 28px */
      "display3": "var(--leo-typography-mobile-size-display3)",
      /** 28px */
      "h1": "var(--leo-typography-mobile-size-h1)",
      /** 22px */
      "h2": "var(--leo-typography-mobile-size-h2)",
      /** 18px */
      "h3": "var(--leo-typography-mobile-size-h3)",
      /** 16px */
      "h4": "var(--leo-typography-mobile-size-h4)",
      /** 10px */
      "label": "var(--leo-typography-mobile-size-label)",
      /** 13px */
      "tableheader": "var(--leo-typography-mobile-size-tableheader)",
      /** 14px */
      "navbutton": "var(--leo-typography-mobile-size-navbutton)",
      /** 12px */
      "buttonSmall": "var(--leo-typography-mobile-size-button-small)",
      /** 13px */
      "buttonDefault": "var(--leo-typography-mobile-size-button-default)",
      /** 15px */
      "buttonLarge": "var(--leo-typography-mobile-size-button-large)",
      /** 16px */
      "buttonJumbo": "var(--leo-typography-mobile-size-button-jumbo)",
      /** 32px */
      "numbers": "var(--leo-typography-mobile-size-numbers)"
    },
    "lineHeight": {
      "toString": [
        ""
      ],
      /** 24px */
      "large": "var(--leo-typography-mobile-line-height-large)",
      /** 22px */
      "default": "var(--leo-typography-mobile-line-height-default)",
      /** 18px */
      "small": "var(--leo-typography-mobile-line-height-small)",
      /** 16px */
      "xSmall": "var(--leo-typography-mobile-line-height-x-small)",
      /** 44px */
      "display1": "var(--leo-typography-mobile-line-height-display1)",
      /** 42px */
      "display2": "var(--leo-typography-mobile-line-height-display2)",
      /** 36px */
      "display3": "var(--leo-typography-mobile-line-height-display3)",
      /** 36px */
      "h1": "var(--leo-typography-mobile-line-height-h1)",
      /** 28px */
      "h2": "var(--leo-typography-mobile-line-height-h2)",
      /** 26px */
      "h3": "var(--leo-typography-mobile-line-height-h3)",
      /** 26px */
      "h4": "var(--leo-typography-mobile-line-height-h4)",
      /** 10px */
      "label": "var(--leo-typography-mobile-line-height-label)",
      /** 16px */
      "tableheader": "var(--leo-typography-mobile-line-height-tableheader)",
      /** 22px */
      "navbutton": "var(--leo-typography-mobile-line-height-navbutton)",
      /** 18px */
      "buttonSmall": "var(--leo-typography-mobile-line-height-button-small)",
      /** 20px */
      "buttonDefault": "var(--leo-typography-mobile-line-height-button-default)",
      /** 22px */
      "buttonLarge": "var(--leo-typography-mobile-line-height-button-large)",
      /** 22px */
      "buttonJumbo": "var(--leo-typography-mobile-line-height-button-jumbo)",
      /** 40px */
      "numbers": "var(--leo-typography-mobile-line-height-numbers)"
    },
    "letterSpacing": {
      "toString": [
        ""
      ],
      /** 0.10000000149011612px */
      "button": "var(--leo-typography-mobile-letter-spacing-button)",
      /** -0.5px */
      "headings": "var(--leo-typography-mobile-letter-spacing-headings)"
    },
    "paragraphSpacing": {
      "toString": [
        ""
      ],
      /** 8px */
      "default": "var(--leo-typography-mobile-paragraph-spacing-default)"
    }
  },
  "ios": {
    "toString": [
      ""
    ],
    "size": {
      "toString": [
        ""
      ],
      /** 17px */
      "large": "var(--leo-typography-ios-size-large)",
      /** 15px */
      "default": "var(--leo-typography-ios-size-default)",
      /** 13px */
      "small": "var(--leo-typography-ios-size-small)",
      /** 11px */
      "xSmall": "var(--leo-typography-ios-size-x-small)",
      /** 44px */
      "display1": "var(--leo-typography-ios-size-display1)",
      /** 36px */
      "display2": "var(--leo-typography-ios-size-display2)",
      /** 28px */
      "display3": "var(--leo-typography-ios-size-display3)",
      /** 32px */
      "h1": "var(--leo-typography-ios-size-h1)",
      /** 28px */
      "h2": "var(--leo-typography-ios-size-h2)",
      /** 22px */
      "h3": "var(--leo-typography-ios-size-h3)",
      /** 16px */
      "h4": "var(--leo-typography-ios-size-h4)",
      /** 10px */
      "label": "var(--leo-typography-ios-size-label)",
      /** 13px */
      "tableheader": "var(--leo-typography-ios-size-tableheader)",
      /** 14px */
      "navbutton": "var(--leo-typography-ios-size-navbutton)",
      /** 13px */
      "buttonSmall": "var(--leo-typography-ios-size-button-small)",
      /** 15px */
      "buttonDefault": "var(--leo-typography-ios-size-button-default)",
      /** 17px */
      "buttonLarge": "var(--leo-typography-ios-size-button-large)",
      /** 18px */
      "buttonJumbo": "var(--leo-typography-ios-size-button-jumbo)",
      /** 32px */
      "numbers": "var(--leo-typography-ios-size-numbers)"
    },
    "lineHeight": {
      "toString": [
        ""
      ],
      /** 24px */
      "large": "var(--leo-typography-ios-line-height-large)",
      /** 22px */
      "default": "var(--leo-typography-ios-line-height-default)",
      /** 18px */
      "small": "var(--leo-typography-ios-line-height-small)",
      /** 16px */
      "xSmall": "var(--leo-typography-ios-line-height-x-small)",
      /** 52px */
      "display1": "var(--leo-typography-ios-line-height-display1)",
      /** 44px */
      "display2": "var(--leo-typography-ios-line-height-display2)",
      /** 40px */
      "display3": "var(--leo-typography-ios-line-height-display3)",
      /** 40px */
      "h1": "var(--leo-typography-ios-line-height-h1)",
      /** 34px */
      "h2": "var(--leo-typography-ios-line-height-h2)",
      /** 28px */
      "h3": "var(--leo-typography-ios-line-height-h3)",
      /** 22px */
      "h4": "var(--leo-typography-ios-line-height-h4)",
      /** 10px */
      "label": "var(--leo-typography-ios-line-height-label)",
      /** 16px */
      "tableheader": "var(--leo-typography-ios-line-height-tableheader)",
      /** 22px */
      "navbutton": "var(--leo-typography-ios-line-height-navbutton)",
      /** 20px */
      "buttonSmall": "var(--leo-typography-ios-line-height-button-small)",
      /** 20px */
      "buttonDefault": "var(--leo-typography-ios-line-height-button-default)",
      /** 22px */
      "buttonLarge": "var(--leo-typography-ios-line-height-button-large)",
      /** 22px */
      "buttonJumbo": "var(--leo-typography-ios-line-height-button-jumbo)",
      /** 40px */
      "numbers": "var(--leo-typography-ios-line-height-numbers)"
    },
    "letterSpacing": {
      "toString": [
        ""
      ],
      /** -0.4000000059604645px */
      "button": "var(--leo-typography-ios-letter-spacing-button)",
      /** -0.4000000059604645px */
      "headings": "var(--leo-typography-ios-letter-spacing-headings)"
    },
    "paragraphSpacing": {
      "toString": [
        ""
      ],
      /** 8px */
      "default": "var(--leo-typography-ios-paragraph-spacing-default)"
    }
  },
  "android": {
    "toString": [
      ""
    ],
    "size": {
      "toString": [
        ""
      ],
      /** 16px */
      "large": "var(--leo-typography-android-size-large)",
      /** 14px */
      "default": "var(--leo-typography-android-size-default)",
      /** 12px */
      "small": "var(--leo-typography-android-size-small)",
      /** 11px */
      "xSmall": "var(--leo-typography-android-size-x-small)",
      /** 45px */
      "display1": "var(--leo-typography-android-size-display1)",
      /** 36px */
      "display2": "var(--leo-typography-android-size-display2)",
      /** 28px */
      "display3": "var(--leo-typography-android-size-display3)",
      /** 32px */
      "h1": "var(--leo-typography-android-size-h1)",
      /** 28px */
      "h2": "var(--leo-typography-android-size-h2)",
      /** 22px */
      "h3": "var(--leo-typography-android-size-h3)",
      /** 16px */
      "h4": "var(--leo-typography-android-size-h4)",
      /** 10px */
      "label": "var(--leo-typography-android-size-label)",
      /** 12px */
      "tableheader": "var(--leo-typography-android-size-tableheader)",
      /** 14px */
      "navbutton": "var(--leo-typography-android-size-navbutton)",
      /** 12px */
      "buttonSmall": "var(--leo-typography-android-size-button-small)",
      /** 14px */
      "buttonDefault": "var(--leo-typography-android-size-button-default)",
      /** 16px */
      "buttonLarge": "var(--leo-typography-android-size-button-large)",
      /** 18px */
      "buttonJumbo": "var(--leo-typography-android-size-button-jumbo)",
      /** 32px */
      "numbers": "var(--leo-typography-android-size-numbers)"
    },
    "lineHeight": {
      "toString": [
        ""
      ],
      /** 24px */
      "large": "var(--leo-typography-android-line-height-large)",
      /** 22px */
      "default": "var(--leo-typography-android-line-height-default)",
      /** 18px */
      "small": "var(--leo-typography-android-line-height-small)",
      /** 16px */
      "xSmall": "var(--leo-typography-android-line-height-x-small)",
      /** 52px */
      "display1": "var(--leo-typography-android-line-height-display1)",
      /** 44px */
      "display2": "var(--leo-typography-android-line-height-display2)",
      /** 40px */
      "display3": "var(--leo-typography-android-line-height-display3)",
      /** 40px */
      "h1": "var(--leo-typography-android-line-height-h1)",
      /** 36px */
      "h2": "var(--leo-typography-android-line-height-h2)",
      /** 28px */
      "h3": "var(--leo-typography-android-line-height-h3)",
      /** 24px */
      "h4": "var(--leo-typography-android-line-height-h4)",
      /** 10px */
      "label": "var(--leo-typography-android-line-height-label)",
      /** 16px */
      "tableheader": "var(--leo-typography-android-line-height-tableheader)",
      /** 22px */
      "navbutton": "var(--leo-typography-android-line-height-navbutton)",
      /** 20px */
      "buttonSmall": "var(--leo-typography-android-line-height-button-small)",
      /** 20px */
      "buttonDefault": "var(--leo-typography-android-line-height-button-default)",
      /** 22px */
      "buttonLarge": "var(--leo-typography-android-line-height-button-large)",
      /** 22px */
      "buttonJumbo": "var(--leo-typography-android-line-height-button-jumbo)",
      /** 40px */
      "numbers": "var(--leo-typography-android-line-height-numbers)"
    },
    "letterSpacing": {
      "toString": [
        ""
      ],
      /** 0.10000000149011612px */
      "button": "var(--leo-typography-android-letter-spacing-button)",
      /** -0.4000000059604645px */
      "headings": "var(--leo-typography-android-letter-spacing-headings)"
    },
    "paragraphSpacing": {
      "toString": [
        ""
      ],
      /** 8px */
      "default": "var(--leo-typography-android-paragraph-spacing-default)"
    }
  },
  "size": {
    "toString": [
      ""
    ],
    /** 16px */
    "large": "var(--leo-typography-size-large)",
    /** 14px */
    "default": "var(--leo-typography-size-default)",
    /** 12px */
    "small": "var(--leo-typography-size-small)",
    /** 11px */
    "xSmall": "var(--leo-typography-size-x-small)",
    /** 46px */
    "display1": "var(--leo-typography-size-display1)",
    /** 40px */
    "display2": "var(--leo-typography-size-display2)",
    /** 34px */
    "display3": "var(--leo-typography-size-display3)",
    /** 32px */
    "h1": "var(--leo-typography-size-h1)",
    /** 28px */
    "h2": "var(--leo-typography-size-h2)",
    /** 22px */
    "h3": "var(--leo-typography-size-h3)",
    /** 16px */
    "h4": "var(--leo-typography-size-h4)",
    /** 10px */
    "label": "var(--leo-typography-size-label)",
    /** 13px */
    "tableheader": "var(--leo-typography-size-tableheader)",
    /** 14px */
    "navbutton": "var(--leo-typography-size-navbutton)",
    /** 12px */
    "buttonSmall": "var(--leo-typography-size-button-small)",
    /** 13px */
    "buttonDefault": "var(--leo-typography-size-button-default)",
    /** 15px */
    "buttonLarge": "var(--leo-typography-size-button-large)",
    /** 16px */
    "buttonJumbo": "var(--leo-typography-size-button-jumbo)",
    /** 36px */
    "numbers": "var(--leo-typography-size-numbers)"
  },
  "lineHeight": {
    "toString": [
      ""
    ],
    /** 24px */
    "large": "var(--leo-typography-line-height-large)",
    /** 22px */
    "default": "var(--leo-typography-line-height-default)",
    /** 18px */
    "small": "var(--leo-typography-line-height-small)",
    /** 16px */
    "xSmall": "var(--leo-typography-line-height-x-small)",
    /** 54px */
    "display1": "var(--leo-typography-line-height-display1)",
    /** 48px */
    "display2": "var(--leo-typography-line-height-display2)",
    /** 40px */
    "display3": "var(--leo-typography-line-height-display3)",
    /** 42px */
    "h1": "var(--leo-typography-line-height-h1)",
    /** 36px */
    "h2": "var(--leo-typography-line-height-h2)",
    /** 28px */
    "h3": "var(--leo-typography-line-height-h3)",
    /** 26px */
    "h4": "var(--leo-typography-line-height-h4)",
    /** 10px */
    "label": "var(--leo-typography-line-height-label)",
    /** 16px */
    "tableheader": "var(--leo-typography-line-height-tableheader)",
    /** 22px */
    "navbutton": "var(--leo-typography-line-height-navbutton)",
    /** 20px */
    "buttonSmall": "var(--leo-typography-line-height-button-small)",
    /** 20px */
    "buttonDefault": "var(--leo-typography-line-height-button-default)",
    /** 22px */
    "buttonLarge": "var(--leo-typography-line-height-button-large)",
    /** 22px */
    "buttonJumbo": "var(--leo-typography-line-height-button-jumbo)",
    /** 46px */
    "numbers": "var(--leo-typography-line-height-numbers)"
  },
  "letterSpacing": {
    "toString": [
      ""
    ],
    /** 0.10000000149011612px */
    "button": "var(--leo-typography-letter-spacing-button)",
    /** -0.5px */
    "headings": "var(--leo-typography-letter-spacing-headings)"
  },
  "paragraphSpacing": {
    "toString": [
      ""
    ],
    /** 8px */
    "default": "var(--leo-typography-paragraph-spacing-default)"
  }
} as const 
export const elevation = {
  "toString": [
    ""
  ],
  /** 1px */
  "xxs": "var(--leo-elevation-xxs)",
  /** 4px */
  "xs": "var(--leo-elevation-xs)",
  /** 8px */
  "s": "var(--leo-elevation-s)",
  /** 16px */
  "m": "var(--leo-elevation-m)",
  /** 24px */
  "l": "var(--leo-elevation-l)",
  /** 36px */
  "xl": "var(--leo-elevation-xl)"
} as const 
export const icon = {
  "toString": [
    ""
  ],
  /** 16px */
  "xs": "var(--leo-icon-xs)",
  /** 18px */
  "s": "var(--leo-icon-s)",
  /** 20px */
  "m": "var(--leo-icon-m)",
  /** 24px */
  "l": "var(--leo-icon-l)",
  /** 32px */
  "xl": "var(--leo-icon-xl)"
} as const 
export const radius = {
  "toString": [
    ""
  ],
  /** 1px */
  "2Xs": "var(--leo-radius-2xs)",
  /** 2px */
  "xs": "var(--leo-radius-xs)",
  /** 4px */
  "s": "var(--leo-radius-s)",
  /** 8px */
  "m": "var(--leo-radius-m)",
  /** 12px */
  "l": "var(--leo-radius-l)",
  /** 16px */
  "xl": "var(--leo-radius-xl)",
  /** 1000px */
  "full": "var(--leo-radius-full)",
  /** 28px */
  "androidDialog": "var(--leo-radius-android-dialog)",
  /** 6px */
  "browserContent": "var(--leo-radius-browser-content)"
} as const 
export const spacing = {
  "toString": [
    ""
  ],
  /** 0px */
  "none": "var(--leo-spacing-none)",
  /** 2px */
  "xs": "var(--leo-spacing-xs)",
  /** 4px */
  "s": "var(--leo-spacing-s)",
  /** 8px */
  "m": "var(--leo-spacing-m)",
  /** 12px */
  "l": "var(--leo-spacing-l)",
  /** 16px */
  "xl": "var(--leo-spacing-xl)",
  /** 24px */
  "2Xl": "var(--leo-spacing-2xl)",
  /** 32px */
  "3Xl": "var(--leo-spacing-3xl)",
  /** 40px */
  "4Xl": "var(--leo-spacing-4xl)",
  /** 48px */
  "5Xl": "var(--leo-spacing-5xl)",
  /** 64px */
  "6Xl": "var(--leo-spacing-6xl)",
  /** 80px */
  "7Xl": "var(--leo-spacing-7xl)",
  /** 96px */
  "8Xl": "var(--leo-spacing-8xl)",
  /** 160px */
  "9Xl": "var(--leo-spacing-9xl)"
} as const
