// @glif/filecoin-wallet-provider/dist/providers/hd-wallet-provider/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HDWalletProvider = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classPrivateFieldGet2 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldGet"));

var _classPrivateFieldSet2 = _interopRequireDefault(require("@babel/runtime/helpers/classPrivateFieldSet"));

var _filecoinAddress = require("@glif/filecoin-address");

var _filecoinMessage = require("@glif/filecoin-message");

var _js = _interopRequireDefault(require("@zondax/filecoin-signing-tools/js"));

var _utils = require("../../utils");

var _errors = require("../../errors");

function _classPrivateFieldInitSpec(obj, privateMap, value) { _checkPrivateRedeclaration(obj, privateMap); privateMap.set(obj, value); }

function _checkPrivateRedeclaration(obj, privateCollection) { if (privateCollection.has(obj)) { throw new TypeError("Cannot initialize the same private elements twice on an object"); } }

var InvalidParamsError = _errors.errors.InvalidParamsError,
    WalletProviderError = _errors.errors.WalletProviderError;

var _seed = /*#__PURE__*/new WeakMap();

var HDWalletProvider = function HDWalletProvider(seed) {
  var _this = this;

  (0, _classCallCheck2["default"])(this, HDWalletProvider);
  (0, _defineProperty2["default"])(this, "type", 'HD_WALLET');
  (0, _defineProperty2["default"])(this, "accountToPath", {});

  _classPrivateFieldInitSpec(this, _seed, {
    writable: true,
    value: void 0
  });

  (0, _defineProperty2["default"])(this, "getAccounts", /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
    var nStart,
        nEnd,
        coinType,
        accounts,
        i,
        path,
        account,
        _args = arguments;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            nStart = _args.length > 0 && _args[0] !== undefined ? _args[0] : 0;
            nEnd = _args.length > 1 && _args[1] !== undefined ? _args[1] : 5;
            coinType = _args.length > 2 && _args[2] !== undefined ? _args[2] : _filecoinAddress.CoinType.MAIN;

            if ((0, _utils.validIndexes)(nStart, nEnd)) {
              _context.next = 5;
              break;
            }

            throw new InvalidParamsError({
              message: 'Invalid indexes provided to getAccounts'
            });

          case 5:
            if (!(coinType !== _filecoinAddress.CoinType.MAIN && coinType !== _filecoinAddress.CoinType.TEST)) {
              _context.next = 7;
              break;
            }

            throw new InvalidParamsError({
              message: 'Invalid coinType passed to getAccounts'
            });

          case 7:
            accounts = [];

            for (i = nStart; i < nEnd; i += 1) {
              path = (0, _utils.createPath)((0, _utils.coinTypeCode)(coinType), i);
              account = _js["default"].keyDerive((0, _classPrivateFieldGet2["default"])(_this, _seed), path, '').address;
              accounts.push(account);
              _this.accountToPath[account] = path;
            }

            return _context.abrupt("return", accounts);

          case 10:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })));
  (0, _defineProperty2["default"])(this, "sign", /*#__PURE__*/function () {
    var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(from, message) {
      var path, msg, _signingTools$keyDeri, private_hexstring, _ref3, signature;

      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              if (!(from !== message.From)) {
                _context2.next = 2;
                break;
              }

              throw new InvalidParamsError({
                message: 'From address mismatch'
              });

            case 2:
              path = _this.accountToPath[from];

              if (path) {
                _context2.next = 5;
                break;
              }

              throw new WalletProviderError({
                message: 'Account was not yet derived from this seed phrase'
              });

            case 5:
              _context2.prev = 5;
              msg = _filecoinMessage.Message.fromLotusType(message);
              _context2.next = 12;
              break;

            case 9:
              _context2.prev = 9;
              _context2.t0 = _context2["catch"](5);
              throw new InvalidParamsError(_context2.t0 instanceof Error ? {
                message: "Invalid message params passed to sign call: ".concat(_context2.t0.message)
              } : undefined);

            case 12:
              _signingTools$keyDeri = _js["default"].keyDerive((0, _classPrivateFieldGet2["default"])(_this, _seed), path, ''), private_hexstring = _signingTools$keyDeri.private_hexstring;
              _ref3 = _js["default"].transactionSign(msg.toZondaxType(), Buffer.from(private_hexstring, 'hex').toString('base64')), signature = _ref3.signature;
              return _context2.abrupt("return", {
                Message: message,
                Signature: {
                  Type: signature.type,
                  Data: signature.data
                }
              });

            case 15:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2, null, [[5, 9]]);
    }));

    return function (_x, _x2) {
      return _ref2.apply(this, arguments);
    };
  }());
  if (!seed) throw new InvalidParamsError();
  (0, _classPrivateFieldSet2["default"])(this, _seed, seed);
};

exports.HDWalletProvider = HDWalletProvider;
