// @brave/brave-ui/old/contentToggleArrow/index.js
import * as React from 'react';
import Separator from '../separator/index';
import { StyledContentToggleArrow, StyledContentToggleArrowControl, StyledContentToggleArrowSummary, StyledContentToggleArrowContent } from './style';
class ContentToggleArrow extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = { open: props.open !== undefined ? props.open : props.defaultOpen };
        this.handleClick = this.handleClick.bind(this);
    }
    componentWillReceiveProps(nextProps) {
        if ('open' in nextProps) {
            this.setState({ open: nextProps.open });
        }
    }
    handleClick(e) {
        this.setState((prevState) => ({ open: !prevState.open }));
        this.props.onClick({
            target: {
                open: this.state.open,
                id: e.target.id
            }
        });
    }
    render() {
        const { id, summary, withSeparator, children } = this.props;
        const { open } = this.state;
        return (React.createElement(StyledContentToggleArrow, { id: id, open: open, withSeparator: withSeparator },
            React.createElement(StyledContentToggleArrowControl, { id: `${id}Control`, open: open, onClick: this.handleClick },
                React.createElement(StyledContentToggleArrowSummary, null, summary)),
            withSeparator && React.createElement(Separator, null),
            React.createElement(StyledContentToggleArrowContent, { open: open, withSeparator: withSeparator }, children),
            open && withSeparator && React.createElement(Separator, null)));
    }
}
export default ContentToggleArrow;
