// @glif/filecoin-address/dist/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Address: true,
  bigintToArray: true,
  getChecksum: true,
  validateChecksum: true,
  newAddress: true,
  newIDAddress: true,
  newActorAddress: true,
  newSecp256k1Address: true,
  newBLSAddress: true,
  decode: true,
  encode: true,
  newFromString: true,
  validateAddressString: true,
  checkAddressString: true,
  idFromAddress: true
};
exports.Address = void 0;
exports.bigintToArray = bigintToArray;
exports.checkAddressString = checkAddressString;
exports.decode = decode;
exports["default"] = void 0;
exports.encode = encode;
exports.getChecksum = getChecksum;
exports.idFromAddress = idFromAddress;
exports.newActorAddress = newActorAddress;
exports.newAddress = newAddress;
exports.newBLSAddress = newBLSAddress;
exports.newFromString = newFromString;
exports.newIDAddress = newIDAddress;
exports.newSecp256k1Address = newSecp256k1Address;
exports.validateAddressString = validateAddressString;
exports.validateChecksum = validateChecksum;

var leb = _interopRequireWildcard(require("leb128"));

var _blakejs = require("blakejs");

var _base = require("./base32");

var uint8arrays = _interopRequireWildcard(require("uint8arrays"));

var _protocol = require("./protocol");

Object.keys(_protocol).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _protocol[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _protocol[key];
    }
  });
});

var _coinType = require("./coinType");

Object.keys(_coinType).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _coinType[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _coinType[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultCoinType = _coinType.CoinType.MAIN;
var base32 = (0, _base.base32)('abcdefghijklmnopqrstuvwxyz234567'); // PayloadHashLength defines the hash length taken over addresses using the
// Actor and SECP256K1 protocols.

var payloadHashLength = 20;

function addressHash(ingest) {
  return (0, _blakejs.blake2b)(ingest, null, payloadHashLength);
}

var Address = /*#__PURE__*/function () {
  function Address(str) {
    var coinType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultCoinType;

    _classCallCheck(this, Address);

    _defineProperty(this, "str", void 0);

    _defineProperty(this, "_protocol", void 0);

    _defineProperty(this, "_coinType", void 0);

    if (!str || str.length < 1) throw new Error('Missing str in address');
    this.str = str;
    this._protocol = this.str[0];

    if (!_protocol.Protocol[this._protocol]) {
      throw new Error("Invalid protocol ".concat(this._protocol));
    }

    this._coinType = coinType;
  }

  _createClass(Address, [{
    key: "network",
    value: function network() {
      return this._coinType;
    }
  }, {
    key: "coinType",
    value: function coinType() {
      return this._coinType;
    }
  }, {
    key: "protocol",
    value: function protocol() {
      return this._protocol;
    }
  }, {
    key: "payload",
    value: function payload() {
      return this.str.slice(1, this.str.length);
    }
    /**
     * toString returns a string representation of this address. If no "coinType"
     * parameter was passed to the constructor the address will be prefixed with
     * the default coinType prefix "f" (mainnet).
     */

  }, {
    key: "toString",
    value: function toString() {
      return encode(this._coinType, this);
    }
    /**
     * equals determines if this address is the "same" address as the passed
     * address. Two addresses are considered equal if they are the same instance
     * OR if their "str" property matches byte for byte.
     */

  }, {
    key: "equals",
    value: function equals(addr) {
      if (this === addr) {
        return true;
      }

      return uint8arrays.equals(this.str, addr.str);
    }
  }]);

  return Address;
}();

exports.Address = Address;

function bigintToArray(v) {
  var tmp = BigInt(v).toString(16);
  if (tmp.length % 2 === 1) tmp = "0".concat(tmp);
  return uint8arrays.fromString(tmp, 'base16');
}

function getChecksum(ingest) {
  return (0, _blakejs.blake2b)(ingest, null, 4);
}

function validateChecksum(ingest, expect) {
  var digest = getChecksum(ingest);
  return uint8arrays.compare(digest, expect);
}

function newAddress(protocol, payload) {
  var coinType = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : defaultCoinType;
  var protocolByte = new Uint8Array([protocol]);
  return new Address(uint8arrays.concat([protocolByte, payload]), coinType);
}

function newIDAddress(id) {
  var coinType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultCoinType;
  return newAddress(_protocol.Protocol.ID, leb.unsigned.encode(id), coinType);
}
/**
 * newActorAddress returns an address using the Actor protocol.
 */


function newActorAddress(data) {
  return newAddress(_protocol.Protocol.ACTOR, addressHash(data));
}
/**
 * newSecp256k1Address returns an address using the SECP256K1 protocol.
 */


function newSecp256k1Address(pubkey) {
  return newAddress(_protocol.Protocol.SECP256K1, addressHash(pubkey));
}
/**
 * newBLSAddress returns an address using the BLS protocol.
 */


function newBLSAddress(pubkey) {
  return newAddress(_protocol.Protocol.BLS, pubkey);
}

function decode(address) {
  checkAddressString(address);
  var coinType = address.slice(0, 1);
  /* tslint:disable-next-line:radix */

  var protocol = parseInt(address.slice(1, 2));
  var raw = address.substring(2, address.length);
  var protocolByte = new Uint8Array([protocol]);

  if (protocol === _protocol.Protocol.ID) {
    return newIDAddress(raw, coinType);
  }

  var payloadChecksum = base32.decode(raw);
  var length = payloadChecksum.length;
  var payload = payloadChecksum.slice(0, length - 4);
  var checksum = payloadChecksum.slice(length - 4, length);

  if (validateChecksum(uint8arrays.concat([protocolByte, payload]), checksum)) {
    throw Error("Checksums don't match");
  }

  var addressObj = newAddress(protocol, payload, coinType);
  if (encode(coinType, addressObj) !== address) throw Error("Did not encode this address properly: ".concat(address));
  return addressObj;
}

function encode(coinType, address) {
  if (!address || !address.str) throw Error('Invalid address');
  var payload = address.payload();

  switch (address.protocol()) {
    case 0:
      {
        return coinType + String(address.protocol()) + leb.unsigned.decode(address.payload());
      }

    default:
      {
        var protocolByte = new Uint8Array([address.protocol()]);
        var checksum = getChecksum(uint8arrays.concat([protocolByte, payload]));
        var bytes = uint8arrays.concat([payload, checksum]);
        return String(coinType) + String(address.protocol()) + base32.encode(bytes);
      }
  }
}

function newFromString(address) {
  return decode(address);
}

function validateAddressString(addressString) {
  try {
    checkAddressString(addressString);
    return true;
  } catch (error) {
    return false;
  }
}

function checkAddressString(address) {
  if (!address) throw Error('No bytes to validate.');
  if (address.length < 3) throw Error('Address is too short to validate.');

  if (address[0] !== 'f' && address[0] !== 't') {
    throw Error('Unknown address coinType.');
  }
  /* tslint:disable-next-line:radix */


  var protocol = parseInt(address[1]);

  switch (protocol) {
    case _protocol.Protocol.ID:
      {
        if (address.length > 22) throw Error('Invalid ID address length.');else if (isNaN(Number(address.slice(2)))) throw Error('Invalid ID address');
        break;
      }

    case _protocol.Protocol.SECP256K1:
      {
        if (address.length !== 41) throw Error('Invalid secp256k1 address length.');
        break;
      }

    case _protocol.Protocol.ACTOR:
      {
        if (address.length !== 41) throw Error('Invalid Actor address length.');
        break;
      }

    case _protocol.Protocol.BLS:
      {
        if (address.length !== 86) throw Error('Invalid BLS address length.');
        break;
      }

    default:
      {
        throw new Error('Invalid address protocol.');
      }
  }
}
/**
 * idFromAddress extracts the ID from an ID address.
 */


function idFromAddress(address) {
  if (address.protocol() !== _protocol.Protocol.ID) throw new Error('Cannot get ID from non ID address'); // An unsigned varint should be less than 2^63 which is < Number.MAX_VALUE.
  // So this number SHOULD be representable in JS and safe to parseInt.
  // https://github.com/multiformats/unsigned-varint
  // TODO: does leb128 enforce the max value?

  return parseInt(leb.unsigned.decode(address.payload()), 10);
}

var _default = {
  Address: Address,
  newAddress: newAddress,
  newIDAddress: newIDAddress,
  newActorAddress: newActorAddress,
  newSecp256k1Address: newSecp256k1Address,
  newBLSAddress: newBLSAddress,
  newFromString: newFromString,
  bigintToArray: bigintToArray,
  decode: decode,
  encode: encode,
  getChecksum: getChecksum,
  validateChecksum: validateChecksum,
  validateAddressString: validateAddressString,
  checkAddressString: checkAddressString,
  idFromAddress: idFromAddress,
  CoinType: _coinType.CoinType,
  Protocol: _protocol.Protocol
};
exports["default"] = _default;
