// @brave/brave-ui/components/buttonsIndicators/button/index.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import { PrimaryButton, SecondaryButton, TertiaryButton, StyledText, StyledIcon } from './style';
export default class ThemedButton extends React.PureComponent {
    getButtonComponent() {
        switch (this.props.level) {
            case 'primary':
                return PrimaryButton;
            case 'secondary':
                return SecondaryButton;
            case 'tertiary':
                return TertiaryButton;
        }
        throw new Error(`Unknown component level: ${this.props.level || '[undefined]'}`);
    }
    render() {
        const _a = this.props, { disabled, className, onClick, id } = _a, deepProps = __rest(_a, ["disabled", "className", "onClick", "id"]);
        const { icon, text } = deepProps;
        if (!onClick && !disabled) {
            console.warn('Button component not provided an onClick handler as a prop');
        }
        let ButtonComponent = this.getButtonComponent();
        return (React.createElement(ButtonComponent, Object.assign({ id: id, className: className, onClick: onClick, disabled: disabled }, deepProps),
            text
                ? React.createElement(StyledText, Object.assign({}, deepProps), text)
                : null,
            icon && icon.image
                ? React.createElement(StyledIcon, Object.assign({}, deepProps), icon.image)
                : null));
    }
}
ThemedButton.displayName = 'Button';
ThemedButton.defaultProps = {
    brand: 'brave',
    size: 'medium',
    type: 'default',
    level: 'primary',
    main: false
};
