// @brave/brave-ui/components/popupModals/modal/index.js
import * as React from 'react';
import { StyledWrapper, StyledDialog, StyledClose, StyledContent } from './style';
import { CloseCircleOIcon } from '../../icons';
export default class Modal extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.outsideClose = () => {
            if (!this.props.outsideClose) {
                return;
            }
            if (this.props.onClose) {
                this.props.onClose();
            }
        };
        this.isLandscape = () => {
            const isLandscape = (window.innerWidth > window.innerHeight);
            return isLandscape;
        };
    }
    render() {
        const { id, onClose, displayCloseButton, children, size, testId, isMobile } = this.props;
        return (React.createElement(StyledWrapper, { id: id, onClick: this.outsideClose, "data-test-id": testId },
            React.createElement(StyledDialog, { size: size },
                displayCloseButton
                    ? React.createElement(StyledClose, { onClick: onClose },
                        React.createElement(CloseCircleOIcon, null))
                    : null,
                React.createElement(StyledContent, { isMobile: isMobile, isLandscape: this.isLandscape() }, children))));
    }
}
Modal.defaultProps = {
    size: 'normal',
    id: 'modal',
    displayCloseButton: true
};
