// @brave/brave-ui/old/page/style.js
import styled from 'styled-components';
import { setTheme } from '../../helpers';
const StyledPage = styled.div `
  font-family: ${(p) => setTheme(p.customStyle, 'fontFamily') || 'inherit'};
  color: ${(p) => setTheme(p.customStyle, 'color') || 'inherit'};
  padding: ${(p) => setTheme(p.customStyle, 'padding') || '40px'};
  background: ${(p) => setTheme(p.customStyle, 'background') || 'inherit'};
  align-items: ${(p) => setTheme(p.customStyle, 'alignItems')};
  justify-content: ${(p) => setTheme(p.customStyle, 'justifyContent')};
  height: ${(p) => setTheme(p.customStyle, 'height') || 'initial'};
  background-size: ${(p) => setTheme(p.customStyle, 'backgroundSize') || 'auto'};
  box-sizing: border-box;
  background-attachment: fixed;
  font-size: inherit;
  display: flex;
  flex-direction: column;
  min-height: 100%;
`;
const StyledPageContent = styled.main `
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  color: inherit;
`;
export { StyledPage, StyledPageContent };
