// @brave/leo/shared/navigationHeader.js
import { S as SvelteComponent, i as init, s as safe_not_equal, a as append_styles, c as create_slot, e as element, b as space, d as attr, f as insert, g as append, u as update_slot_base, m as get_all_dirty_from_scope, n as get_slot_changes, k as transition_in, h as transition_out, o as detach } from './a4e1d370.js';

/* src/components/navigation/navigationHeader.svelte generated by Svelte v4.2.19 */

function add_css(target) {
	append_styles(target, "svelte-33ky3s", ".leo-navigation-header.svelte-33ky3s{display:flex;gap:var(--leo-spacing-m);align-items:center;justify-self:self-start;padding:var(--leo-navigation-header-padding, var(--leo-spacing-2xl)) 20px}");
}

const get_logo_slot_changes = dirty => ({});
const get_logo_slot_context = ctx => ({});

function create_fragment(ctx) {
	let header;
	let t;
	let current;
	const logo_slot_template = /*#slots*/ ctx[1].logo;
	const logo_slot = create_slot(logo_slot_template, ctx, /*$$scope*/ ctx[0], get_logo_slot_context);
	const default_slot_template = /*#slots*/ ctx[1].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[0], null);

	return {
		c() {
			header = element("header");
			if (logo_slot) logo_slot.c();
			t = space();
			if (default_slot) default_slot.c();
			attr(header, "class", "leo-navigation-header svelte-33ky3s");
		},
		m(target, anchor) {
			insert(target, header, anchor);

			if (logo_slot) {
				logo_slot.m(header, null);
			}

			append(header, t);

			if (default_slot) {
				default_slot.m(header, null);
			}

			current = true;
		},
		p(ctx, [dirty]) {
			if (logo_slot) {
				if (logo_slot.p && (!current || dirty & /*$$scope*/ 1)) {
					update_slot_base(
						logo_slot,
						logo_slot_template,
						ctx,
						/*$$scope*/ ctx[0],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[0])
						: get_slot_changes(logo_slot_template, /*$$scope*/ ctx[0], dirty, get_logo_slot_changes),
						get_logo_slot_context
					);
				}
			}

			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 1)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[0],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[0])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[0], dirty, null),
						null
					);
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(logo_slot, local);
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(logo_slot, local);
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (detaching) {
				detach(header);
			}

			if (logo_slot) logo_slot.d(detaching);
			if (default_slot) default_slot.d(detaching);
		}
	};
}

function instance($$self, $$props, $$invalidate) {
	let { $$slots: slots = {}, $$scope } = $$props;

	$$self.$$set = $$props => {
		if ('$$scope' in $$props) $$invalidate(0, $$scope = $$props.$$scope);
	};

	return [$$scope, slots];
}

class NavigationHeader extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance, create_fragment, safe_not_equal, {}, add_css);
	}
}

export { NavigationHeader as default };
