// @brave/leo/shared/navigationMenu.js
import { S as SvelteComponent, i as init, s as safe_not_equal, a as append_styles, c as create_slot, e as element, d as attr, t as toggle_class, f as insert, u as update_slot_base, m as get_all_dirty_from_scope, n as get_slot_changes, k as transition_in, h as transition_out, o as detach, P as assign, T as exclude_internal_props } from './a4e1d370.js';

/* src/components/navigation/navigationMenu.svelte generated by Svelte v4.2.19 */

function add_css(target) {
	append_styles(target, "svelte-3dd5zt", "ul.svelte-3dd5zt{display:flex;flex-direction:var(--nav-direction);gap:var(--leo-spacing-m);margin:0;padding:var(--leo-spacing-2xl) 0;list-style:none}ul.isSubnav.svelte-3dd5zt,:host([slot='subnav']) ul.svelte-3dd5zt{padding:var(--leo-spacing-m) 0 0;margin-left:33px;border-left:1px solid var(--leo-color-divider-subtle)}[hidden].svelte-3dd5zt{display:none !important}");
}

function create_fragment(ctx) {
	let ul;
	let ul_hidden_value;
	let current;
	const default_slot_template = /*#slots*/ ctx[3].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[2], null);

	return {
		c() {
			ul = element("ul");
			if (default_slot) default_slot.c();
			ul.hidden = ul_hidden_value = /*$$props*/ ctx[1].hidden ? true : undefined;
			attr(ul, "class", "svelte-3dd5zt");
			toggle_class(ul, "isSubnav", /*isSubnav*/ ctx[0]);
		},
		m(target, anchor) {
			insert(target, ul, anchor);

			if (default_slot) {
				default_slot.m(ul, null);
			}

			current = true;
		},
		p(ctx, [dirty]) {
			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 4)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[2],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[2])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[2], dirty, null),
						null
					);
				}
			}

			if (!current || dirty & /*$$props*/ 2 && ul_hidden_value !== (ul_hidden_value = /*$$props*/ ctx[1].hidden ? true : undefined)) {
				ul.hidden = ul_hidden_value;
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (detaching) {
				detach(ul);
			}

			if (default_slot) default_slot.d(detaching);
		}
	};
}

function instance($$self, $$props, $$invalidate) {
	let { $$slots: slots = {}, $$scope } = $$props;
	let isSubnav = $$props.slot === 'subnav';

	$$self.$$set = $$new_props => {
		$$invalidate(1, $$props = assign(assign({}, $$props), exclude_internal_props($$new_props)));
		if ('$$scope' in $$new_props) $$invalidate(2, $$scope = $$new_props.$$scope);
	};

	$$props = exclude_internal_props($$props);
	return [isSubnav, $$props, $$scope, slots];
}

class NavigationMenu extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance, create_fragment, safe_not_equal, {}, add_css);
	}
}

export { NavigationMenu as default };
