// @brave/leo/shared/progressBar.js
import { S as SvelteComponent, i as init, s as safe_not_equal, a as append_styles, e as element, d as attr, C as set_style, f as insert, g as append, z as noop, o as detach } from './a4e1d370.js';

/* src/components/progress/progressBar.svelte generated by Svelte v4.2.19 */

function add_css(target) {
	append_styles(target, "svelte-1ws2tl8", ":host{display:block}.leo-progressbar.svelte-1ws2tl8.svelte-1ws2tl8{--radius:var(--leo-progressbar-radius, 10px);--transition-duration:var(--leo-progressbar-transition-duration, 0.2s);background:var(--leo-progressbar-background-color, var(--leo-color-primary-20));height:var(--leo-progressbar-height, 4px);border-radius:var(--radius)}.leo-progressbar.svelte-1ws2tl8 .progress.svelte-1ws2tl8{height:100%;width:100%;transition:width var(--transition-duration) ease-in-out;background:var(--leo-progressbar-color, var(--leo-color-button-background));border-radius:var(--radius)}");
}

function create_fragment(ctx) {
	let div1;
	let div0;

	return {
		c() {
			div1 = element("div");
			div0 = element("div");
			attr(div0, "class", "progress svelte-1ws2tl8");
			set_style(div0, "width", /*progress*/ ctx[0] * 100 + "%");
			attr(div1, "class", "leo-progressbar svelte-1ws2tl8");
		},
		m(target, anchor) {
			insert(target, div1, anchor);
			append(div1, div0);
		},
		p(ctx, [dirty]) {
			if (dirty & /*progress*/ 1) {
				set_style(div0, "width", /*progress*/ ctx[0] * 100 + "%");
			}
		},
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) {
				detach(div1);
			}
		}
	};
}

function instance($$self, $$props, $$invalidate) {
	let { progress = 0 } = $$props;

	$$self.$$set = $$props => {
		if ('progress' in $$props) $$invalidate(0, progress = $$props.progress);
	};

	return [progress];
}

class ProgressBar extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance, create_fragment, safe_not_equal, { progress: 0 }, add_css);
	}
}

export { ProgressBar as default };
