// @brave/brave-ui/old/switchButton/style.js
import styled, { css } from 'styled-components';
import defaultTheme from '../defaultTheme';
import { setValueBasedOnSize, setTheme } from '../../helpers';
const StyledSwitchButtonWrapper = styled.div `
  box-sizing: border-box;
  display: flex;
  align-items: center;
  font-family: inherit;
  color: inherit;

  ${(p) => p.disabled && css `
    opacity: 0.3;
    user-select: none;`}
`;
const StyledSwitchButtonLabel = styled.label `
  box-sizing: border-box;
  font-size: ${(p) => setValueBasedOnSize(p.switchSize, 'inherit', 'inherit', '24px')};
  margin-left: ${(p) => setValueBasedOnSize(p.switchSize, '10px', '10px', '18px')};
  margin-right: ${(p) => setValueBasedOnSize(p.switchSize, '10px', '10px', '18px')};
  font-weight: 400;
  font-family: inherit;
  color: ${(p) => setTheme(p.customStyle, 'labelColor') || 'inherit'};
  user-select: ${(p) => setTheme(p.customStyle, 'userSelect') || 'none'};
`;
const StyledSwitchButton = styled.input `
  box-sizing: border-box;
  appearance: none;
  position: relative;
  min-width: ${(p) => setValueBasedOnSize(p.switchSize, '48px', '40px', '60px')};
  height: ${(p) => setValueBasedOnSize(p.switchSize, '16px', '14px', '26px')};
  border-radius: ${(p) => setValueBasedOnSize(p.switchSize, '16px', '14px', '26px')};
  border: 2px solid ${defaultTheme.switchButton.uncheckedBg};
  padding: 0 2px;
  background-color: white;
  outline: none;
  transition: linear 0.2s;
  box-shadow:
    inset
    -${(p) => setValueBasedOnSize(p.switchSize, '31px', '25px', '33px')} 0px 0px 0px
    ${defaultTheme.switchButton.uncheckedBg};

  &:checked {
    border: 2px solid ${defaultTheme.switchButton.checkedBg};
    box-shadow:
      inset
      ${(p) => setValueBasedOnSize(p.switchSize, '31px', '25px', '33px')} 0px 0px 0px
      ${defaultTheme.switchButton.checkedBg};
  }
`;
export { StyledSwitchButtonWrapper, StyledSwitchButtonLabel, StyledSwitchButton };
