// @glif/filecoin-wallet-provider/lib/errors/errors.js
import _extends from "@babel/runtime/helpers/extends";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _wrapNativeSuper from "@babel/runtime/helpers/wrapNativeSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/* tslint:disable max-classes-per-file */
var WalletProviderError = /*#__PURE__*/function (_Error) {
  _inherits(WalletProviderError, _Error);

  var _super = _createSuper(WalletProviderError);

  function WalletProviderError(args) {
    var _this;

    _classCallCheck(this, WalletProviderError);

    _this = _super.call(this, args.message);

    _defineProperty(_assertThisInitialized(_this), "code", void 0);

    Object.setPrototypeOf(_assertThisInitialized(_this), WalletProviderError.prototype);
    if (args.code) _this.code = args.code;
    return _this;
  }

  _createClass(WalletProviderError, [{
    key: "toJSON",
    value: function toJSON() {
      var res = {
        message: this.message
      };
      if (this.code) res.code = this.code;
      return res;
    }
  }]);

  return WalletProviderError;
}( /*#__PURE__*/_wrapNativeSuper(Error));

var LedgerLostConnectionError = /*#__PURE__*/function (_WalletProviderError) {
  _inherits(LedgerLostConnectionError, _WalletProviderError);

  var _super2 = _createSuper(LedgerLostConnectionError);

  function LedgerLostConnectionError() {
    var _this2;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref);

    _classCallCheck(this, LedgerLostConnectionError);

    _this2 = _super2.call(this, _objectSpread({
      message: args.message || 'connection lost'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this2), LedgerLostConnectionError.prototype);
    return _this2;
  }

  return LedgerLostConnectionError;
}(WalletProviderError);

var TransportNotSupportedError = /*#__PURE__*/function (_WalletProviderError2) {
  _inherits(TransportNotSupportedError, _WalletProviderError2);

  var _super3 = _createSuper(TransportNotSupportedError);

  function TransportNotSupportedError() {
    var _this3;

    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref2);

    _classCallCheck(this, TransportNotSupportedError);

    _this3 = _super3.call(this, _objectSpread({
      message: args.message || 'transport not supported by device'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this3), TransportNotSupportedError.prototype);
    return _this3;
  }

  return TransportNotSupportedError;
}(WalletProviderError);

var LedgerReplugError = /*#__PURE__*/function (_WalletProviderError3) {
  _inherits(LedgerReplugError, _WalletProviderError3);

  var _super4 = _createSuper(LedgerReplugError);

  function LedgerReplugError() {
    var _this4;

    var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref3);

    _classCallCheck(this, LedgerReplugError);

    _this4 = _super4.call(this, _objectSpread({
      message: args.message || 'unknown'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this4), LedgerReplugError.prototype);
    return _this4;
  }

  return LedgerReplugError;
}(WalletProviderError);

var LedgerDisconnectedError = /*#__PURE__*/function (_WalletProviderError4) {
  _inherits(LedgerDisconnectedError, _WalletProviderError4);

  var _super5 = _createSuper(LedgerDisconnectedError);

  function LedgerDisconnectedError() {
    var _this5;

    var _ref4 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref4);

    _classCallCheck(this, LedgerDisconnectedError);

    _this5 = _super5.call(this, _objectSpread({
      message: args.message || 'ledger device disconnected'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this5), LedgerDisconnectedError.prototype);
    return _this5;
  }

  return LedgerDisconnectedError;
}(WalletProviderError);

var LedgerInUseByAnotherApp = /*#__PURE__*/function (_WalletProviderError5) {
  _inherits(LedgerInUseByAnotherApp, _WalletProviderError5);

  var _super6 = _createSuper(LedgerInUseByAnotherApp);

  function LedgerInUseByAnotherApp() {
    var _this6;

    var _ref5 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref5);

    _classCallCheck(this, LedgerInUseByAnotherApp);

    _this6 = _super6.call(this, _objectSpread({
      message: args.message || 'ledger device in use by another app'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this6), LedgerInUseByAnotherApp.prototype);
    return _this6;
  }

  return LedgerInUseByAnotherApp;
}(WalletProviderError);

var LedgerNotFoundError = /*#__PURE__*/function (_WalletProviderError6) {
  _inherits(LedgerNotFoundError, _WalletProviderError6);

  var _super7 = _createSuper(LedgerNotFoundError);

  function LedgerNotFoundError() {
    var _this7;

    var _ref6 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref6);

    _classCallCheck(this, LedgerNotFoundError);

    _this7 = _super7.call(this, _objectSpread({
      message: args.message || 'ledger device not found'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this7), LedgerNotFoundError.prototype);
    return _this7;
  }

  return LedgerNotFoundError;
}(WalletProviderError);

var LedgerDeviceLockedError = /*#__PURE__*/function (_WalletProviderError7) {
  _inherits(LedgerDeviceLockedError, _WalletProviderError7);

  var _super8 = _createSuper(LedgerDeviceLockedError);

  function LedgerDeviceLockedError() {
    var _this8;

    var _ref7 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref7);

    _classCallCheck(this, LedgerDeviceLockedError);

    _this8 = _super8.call(this, _objectSpread({
      message: args.message || 'ledger device locked'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this8), LedgerDeviceLockedError.prototype);
    return _this8;
  }

  return LedgerDeviceLockedError;
}(WalletProviderError);

var LedgerFilecoinAppBadVersionError = /*#__PURE__*/function (_WalletProviderError8) {
  _inherits(LedgerFilecoinAppBadVersionError, _WalletProviderError8);

  var _super9 = _createSuper(LedgerFilecoinAppBadVersionError);

  function LedgerFilecoinAppBadVersionError() {
    var _this9;

    var _ref8 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref8);

    _classCallCheck(this, LedgerFilecoinAppBadVersionError);

    _this9 = _super9.call(this, _objectSpread({
      message: args.message || 'bad filecoin app version on ledger device'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this9), LedgerFilecoinAppBadVersionError.prototype);
    return _this9;
  }

  return LedgerFilecoinAppBadVersionError;
}(WalletProviderError);

var LedgerFilecoinAppNotOpenError = /*#__PURE__*/function (_WalletProviderError9) {
  _inherits(LedgerFilecoinAppNotOpenError, _WalletProviderError9);

  var _super10 = _createSuper(LedgerFilecoinAppNotOpenError);

  function LedgerFilecoinAppNotOpenError() {
    var _this10;

    var _ref9 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref9);

    _classCallCheck(this, LedgerFilecoinAppNotOpenError);

    _this10 = _super10.call(this, _objectSpread({
      message: args.message || 'filecoin app not open on device'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this10), LedgerFilecoinAppNotOpenError.prototype);
    return _this10;
  }

  return LedgerFilecoinAppNotOpenError;
}(WalletProviderError);

var LedgerDeviceBusyError = /*#__PURE__*/function (_WalletProviderError10) {
  _inherits(LedgerDeviceBusyError, _WalletProviderError10);

  var _super11 = _createSuper(LedgerDeviceBusyError);

  function LedgerDeviceBusyError() {
    var _this11;

    var _ref10 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref10);

    _classCallCheck(this, LedgerDeviceBusyError);

    _this11 = _super11.call(this, _objectSpread({
      message: args.message || 'ledger device locked or busy'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this11), LedgerDeviceBusyError.prototype);
    return _this11;
  }

  return LedgerDeviceBusyError;
}(WalletProviderError);

var InvalidParamsError = /*#__PURE__*/function (_WalletProviderError11) {
  _inherits(InvalidParamsError, _WalletProviderError11);

  var _super12 = _createSuper(InvalidParamsError);

  function InvalidParamsError() {
    var _this12;

    var _ref11 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref11);

    _classCallCheck(this, InvalidParamsError);

    _this12 = _super12.call(this, _objectSpread({
      message: args.message || 'invalid params'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this12), InvalidParamsError.prototype);
    return _this12;
  }

  return InvalidParamsError;
}(WalletProviderError);

var UnsupportedKeyTypeError = /*#__PURE__*/function (_WalletProviderError12) {
  _inherits(UnsupportedKeyTypeError, _WalletProviderError12);

  var _super13 = _createSuper(UnsupportedKeyTypeError);

  function UnsupportedKeyTypeError() {
    var _this13;

    var _ref12 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref12);

    _classCallCheck(this, UnsupportedKeyTypeError);

    _this13 = _super13.call(this, _objectSpread({
      message: args.message || 'unsupported key type'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this13), UnsupportedKeyTypeError.prototype);
    return _this13;
  }

  return UnsupportedKeyTypeError;
}(WalletProviderError);

var TransactionRejectedError = /*#__PURE__*/function (_WalletProviderError13) {
  _inherits(TransactionRejectedError, _WalletProviderError13);

  var _super14 = _createSuper(TransactionRejectedError);

  function TransactionRejectedError() {
    var _this14;

    var _ref13 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref13);

    _classCallCheck(this, TransactionRejectedError);

    _this14 = _super14.call(this, _objectSpread({
      message: args.message || 'Transaction rejected'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this14), TransactionRejectedError.prototype);
    return _this14;
  }

  return TransactionRejectedError;
}(WalletProviderError);

var MetaMaskError = /*#__PURE__*/function (_WalletProviderError14) {
  _inherits(MetaMaskError, _WalletProviderError14);

  var _super15 = _createSuper(MetaMaskError);

  function MetaMaskError() {
    var _this15;

    var _ref14 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref14);

    _classCallCheck(this, MetaMaskError);

    _this15 = _super15.call(this, _objectSpread({
      message: args.message || 'MetaMask error'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this15), MetaMaskError.prototype);
    return _this15;
  }

  return MetaMaskError;
}(WalletProviderError);

var MetaMaskNotInstalledError = /*#__PURE__*/function (_WalletProviderError15) {
  _inherits(MetaMaskNotInstalledError, _WalletProviderError15);

  var _super16 = _createSuper(MetaMaskNotInstalledError);

  function MetaMaskNotInstalledError() {
    var _this16;

    var _ref15 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref15);

    _classCallCheck(this, MetaMaskNotInstalledError);

    _this16 = _super16.call(this, _objectSpread({
      message: args.message || 'MetaMask not installed'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this16), MetaMaskNotInstalledError.prototype);
    return _this16;
  }

  return MetaMaskNotInstalledError;
}(WalletProviderError);

var MetaMaskSnapsNotSupportedError = /*#__PURE__*/function (_WalletProviderError16) {
  _inherits(MetaMaskSnapsNotSupportedError, _WalletProviderError16);

  var _super17 = _createSuper(MetaMaskSnapsNotSupportedError);

  function MetaMaskSnapsNotSupportedError() {
    var _this17;

    var _ref16 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref16);

    _classCallCheck(this, MetaMaskSnapsNotSupportedError);

    _this17 = _super17.call(this, _objectSpread({
      message: args.message || 'The installed version of MetaMask does not support Snaps'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this17), MetaMaskSnapsNotSupportedError.prototype);
    return _this17;
  }

  return MetaMaskSnapsNotSupportedError;
}(WalletProviderError);

var MetaMaskLockedError = /*#__PURE__*/function (_WalletProviderError17) {
  _inherits(MetaMaskLockedError, _WalletProviderError17);

  var _super18 = _createSuper(MetaMaskLockedError);

  function MetaMaskLockedError() {
    var _this18;

    var _ref17 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref17);

    _classCallCheck(this, MetaMaskLockedError);

    _this18 = _super18.call(this, _objectSpread({
      message: args.message || 'MetaMask locked'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this18), MetaMaskLockedError.prototype);
    return _this18;
  }

  return MetaMaskLockedError;
}(WalletProviderError);

var MetaMaskFilSnapNotInstalledError = /*#__PURE__*/function (_WalletProviderError18) {
  _inherits(MetaMaskFilSnapNotInstalledError, _WalletProviderError18);

  var _super19 = _createSuper(MetaMaskFilSnapNotInstalledError);

  function MetaMaskFilSnapNotInstalledError() {
    var _this19;

    var _ref18 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    var args = _extends({}, _ref18);

    _classCallCheck(this, MetaMaskFilSnapNotInstalledError);

    _this19 = _super19.call(this, _objectSpread({
      message: args.message || 'FILSnap not installed'
    }, args));
    Object.setPrototypeOf(_assertThisInitialized(_this19), MetaMaskFilSnapNotInstalledError.prototype);
    return _this19;
  }

  return MetaMaskFilSnapNotInstalledError;
}(WalletProviderError);

export default {
  InvalidParamsError: InvalidParamsError,
  LedgerLostConnectionError: LedgerLostConnectionError,
  LedgerReplugError: LedgerReplugError,
  LedgerDisconnectedError: LedgerDisconnectedError,
  LedgerInUseByAnotherApp: LedgerInUseByAnotherApp,
  LedgerNotFoundError: LedgerNotFoundError,
  LedgerDeviceLockedError: LedgerDeviceLockedError,
  LedgerFilecoinAppBadVersionError: LedgerFilecoinAppBadVersionError,
  LedgerFilecoinAppNotOpenError: LedgerFilecoinAppNotOpenError,
  LedgerDeviceBusyError: LedgerDeviceBusyError,
  MetaMaskError: MetaMaskError,
  MetaMaskNotInstalledError: MetaMaskNotInstalledError,
  MetaMaskSnapsNotSupportedError: MetaMaskSnapsNotSupportedError,
  MetaMaskLockedError: MetaMaskLockedError,
  MetaMaskFilSnapNotInstalledError: MetaMaskFilSnapNotInstalledError,
  TransactionRejectedError: TransactionRejectedError,
  TransportNotSupportedError: TransportNotSupportedError,
  UnsupportedKeyTypeError: UnsupportedKeyTypeError,
  WalletProviderError: WalletProviderError
};
